// @sap/cds-compiler/lib/checks/onConditions.js
'use strict';

const { forEachGeneric, isBuiltinType } = require('../model/csnUtils');

// Only to be used with validator.js - a correct this value needs to be provided!

/**
 * Prepare the ref steps so that they are loggable
 *
 * @param {any} refStep part of a ref
 * @returns {string} Loggable string
 */
function logReady(refStep) {
  return refStep.id || refStep;
}

/**
 * Check that the other side of the comparison is a valid $self backlink
 *
 * - operator "="
 * - nothing but "$self", no further steps
 *
 * @param {Array} on On-Condition
 * @param {number} startIndex Index of the current expression to "look around"
 * @returns {boolean} True if valid
 */
function otherSideIsValidDollarSelf(on, startIndex) {
  if (on[startIndex - 1] && on[startIndex - 1] === '=') {
    if (on[startIndex - 2]) {
      const { ref } = on[startIndex - 2];
      return ref && ref.length === 1 && ( ref[0] === '$self' || ref[0] === '$projection' );
    }
    return false;
  }
  else if (on[startIndex + 1] && on[startIndex + 1] === '=') {
    if (on[startIndex + 2]) {
      const { ref } = on[startIndex + 2];
      return ref && ref.length === 1 && ( ref[0] === '$self' || ref[0] === '$projection' );
    }
    return false;
  }
  return false;
}

/**
 * Check that the opposite operand to a relational term is something
 * structured that can be used for tuple expansion. This can either be a
 * real 'elements' thing or a managed association/composition with foreign keys.
 *
 * @param {Array} on the on condition which to check
 * @param {number} startIndex the index of the relational term in the on condition array
 * @returns {boolean} indicates whether the other side of a relational term is expandable
 */
function otherSideIsExpandableStructure(on, startIndex) {
  if (on[startIndex - 1] && [ '=', '<', '>', '>=', '<=', '!=', '<>' ].includes(on[startIndex - 1]))
    return isOk(resolveArtifactType.call(this, on[startIndex - 2]._art));

  else if (on[startIndex + 1] && [ '=', '<', '>', '>=', '<=', '!=', '<>' ].includes(on[startIndex + 1]))
    return isOk(resolveArtifactType.call(this, on[startIndex + 2]._art));

  return false;

  /**
   * Artifact is structured or a managed association/compoisition
   *
   * @param {CSN.Artifact} art Artifact
   * @returns {boolean} True if expandable
   */
  function isOk(art) {
    return !!(art && (art.elements || (art.target && art.keys)));
  }
}

/**
 * Get the real type of an artifact
 *
 * @param {object} art Whatever _art by csnRefs can be - element or artifact
 * @returns {object} final artifact type
 */
function resolveArtifactType(art) {
  if (art && art.type && !isBuiltinType(art.type))
    return this.getFinalBaseType(art);

  return art;
}

/**
 * Validate an on-condition
 *
 * - no traversal of unmanaged associations
 * - only use managed associations to access their foreign keys
 * - no filters
 * - no parameters
 * - must end in scalar type - unless $self comparison
 *
 * @param {object} member Member
 * @param {string} memberName Name of the member
 * @param {string} property Current property (part of forEachMember)
 * @param {CSN.Path} path CSN Path to current member
 */
function validateOnCondition(member, memberName, property, path) {
  if (member && member.on) {
    for (let i = 0; i < member.on.length; i++) {
      if (member.on[i].ref) {
        const { ref } = member.on[i];
        // eslint-disable-next-line prefer-const
        let { _links, _art, $scope } = member.on[i];
        if (!_links)
          continue;
        const validDollarSelf = otherSideIsValidDollarSelf(member.on, i);
        const validStructuredElement = otherSideIsExpandableStructure.call(this, member.on, i);
        for (let j = 0; j < _links.length - 1; j++) {
          const csnPath = path.concat([ 'on', i, 'ref', j ]);

          // For error messages
          const id = logReady(ref[j]);
          const target = ref.map(ps => logReady(ps)).join('.');

          if (_links[j].art.target && !((_links[j].art === member) || ref[j] === '$self' || ref[j] === '$projection' || (validDollarSelf && j === _links.length - 1))) {
            if (_links[j].art.on) {
              // It's an unmanaged association - traversal is always forbidden
              this.error(null, csnPath, { id, target }, 'ON-Conditions can\'t follow unmanaged associations, step $(ID) of path $(TARGET)');
            }
            else {
              // It's a managed association - access of the foreign keys is allowed
              const nextRef = ref[j + 1].id || ref[j + 1];
              if (!_links[j].art.keys.some(r => r.ref[0] === nextRef))
                this.error(null, csnPath, { id, target }, 'ON-Conditions can only follow managed associations to the foreign keys of the managed association, step $(ID) of path $(TARGET)');
            }
          }

          if (_links[j].art.virtual)
            this.error(null, csnPath, { id, target }, 'Virtual elements can\'t be used in ON-Conditions, step $(ID) of path $(TARGET)');

          if (ref[j].where)
            this.error(null, csnPath, { id, target }, 'ON-Conditions must not contain filters, step $(ID) of path $(TARGET)');

          if (ref[j].args)
            this.error(null, csnPath, { id, target }, 'ON-Conditions must not contain parameters, step $(ID) of path $(TARGET)');
        }
        if (_art && $scope !== '$self') {
          _art = resolveArtifactType.call(this, _art);
          // For error messages
          const target = ref.map(ps => logReady(ps)).join('.');
          const onPath = path.concat([ 'on', i, 'ref', ref.length - 1 ]);
          // Paths of an ON condition may end on a structured element or an association only if:
          // 1) Both operands in the expression end on a structured element or on
          //    a managed association (that are both expandable)
          // 2) Path ends on an association (managed or unmanaged) and the other operand is a '$self'

          // If this path ends structured or on an association, perform the check:
          if ((_art.elements || _art.target) &&
             !( /* 1) */ (_art.elements || _art.target && _art.keys) && validStructuredElement ||
             /* 2) */ (_art.target && validDollarSelf)) &&
             !_art.virtual) {
            this.error(null, onPath, { target },
                       'The last path of an on-condition must be a scalar value, path $(TARGET)');
          }
          else if (_art.items && !_art.virtual) {
            this.error(null, onPath, { target },
                       'ON-Conditions can\'t use array-like elements, path $(TARGET)');
          }
          else if (_art.virtual) {
            this.error(null, onPath, { target },
                       'Virtual elements can\'t be used in ON-Conditions, path $(TARGET)');
          }
        }
      }
    }
  }
}

/**
 * Run the above validations also for mixins.
 *
 * @param {CSN.Query} query query object
 * @param {CSN.Path} path path to the query
 */
function validateMixinOnCondition(query, path) {
  if (query.SELECT && query.SELECT.mixin)
    forEachGeneric( query.SELECT, 'mixin', validateOnCondition.bind(this), path );
}
module.exports = { validateOnCondition, validateMixinOnCondition };
