// @sap/cds-runtime/lib/fiori/generic/new.js
const cds = require('../../cds')
const { INSERT, SELECT, UPDATE } = cds.ql

const onDraftActivate = require('./activate')._handler
const { isNavigationToMany } = require('../../cds-services/services/utils/compositionTree')
const { removeDraftUUID, ensureDraftsSuffix } = require('../utils/handler')
const { DRAFT_COLUMNS } = require('../../common/constants/draft')
const { getVirtuals, postProcessVirtuals } = require('../../db/generic/virtual')

const _getUpdateDraftAdminCQN = ({ user }, draftUUID) => {
  return UPDATE('DRAFT.DraftAdministrativeData')
    .data({
      InProcessByUser: user.id,
      LastChangedByUser: user.id,
      LastChangeDateTime: new Date().toISOString()
    })
    .where({ DraftUUID: draftUUID })
}

const _getInsertDraftAdminCQN = ({ user }, uuid) => {
  const time = new Date().toISOString()

  return INSERT.into('DRAFT.DraftAdministrativeData').entries({
    DraftUUID: uuid,
    CreationDateTime: time,
    CreatedByUser: user.id,
    LastChangeDateTime: time,
    LastChangedByUser: user.id,
    DraftIsCreatedByMe: true,
    DraftIsProcessedByMe: true,
    InProcessByUser: user.id
  })
}

const _getInsertDataCQN = (context, draftUUID) => {
  const draftName = ensureDraftsSuffix(context.target.name)

  const insertData = INSERT.into(draftName).entries(context.query.INSERT.entries[0]) // entries is always set because there are no entities without keys

  context.data.IsActiveEntity = false
  context.data.HasDraftEntity = false
  context.data.HasActiveEntity = false
  context.data.DraftAdministrativeData_DraftUUID = draftUUID

  return insertData
}

/**
 * Generic Handler for CREATE requests in the context of draft.
 * In case of success it returns the created entry.
 *
 * @param req
 * @param next
 */
const _handler = function (req, next) {
  if (!req._draftMetadata) {
    // REVISIT: when is this the case?
    return onDraftActivate(req, next)
  }

  // fill default values
  const elements = req.target.elements
  for (const column in elements) {
    const col = elements[column]
    if (col.default !== undefined && !DRAFT_COLUMNS.includes(column)) {
      req.data[col.name] = 'val' in col.default ? col.default.val : col.default
    }
  }

  const navigationToMany = isNavigationToMany(req)

  const adminDataCQN = navigationToMany
    ? _getUpdateDraftAdminCQN(req, req.data.DraftAdministrativeData_DraftUUID)
    : _getInsertDraftAdminCQN(req, req.data.DraftAdministrativeData_DraftUUID)
  const insertDataCQN = _getInsertDataCQN(req, req.data.DraftAdministrativeData_DraftUUID)

  // read data as on db and return
  const keyName = Object.values(req.target.elements).find(val => val.key).name
  const readInsertDataCQN = SELECT.from(insertDataCQN.INSERT.into).where(keyName, '=', req.data[keyName])
  const virtuals = getVirtuals({ query: readInsertDataCQN, target: req.target }, this.model)
  const dbtx = cds.tx(req)

  return Promise.all([dbtx.run(adminDataCQN), dbtx.run(insertDataCQN)])
    .then(() => {
      return dbtx.run(readInsertDataCQN)
    })
    .then(result => {
      if (result.length === 0) {
        req.reject(404)
      }
      postProcessVirtuals(virtuals, result)
      return removeDraftUUID(result[0])
    })
}

const { ODATA, COMMON } = require('../../common/constants/annotation')
const _relevant = e => e[ODATA.DRAFT] || e[COMMON.DRAFT_NODE.PREP_ACTION]

module.exports = function () {
  for (const entity of Object.values(this.entities).filter(_relevant)) {
    this.on('NEW', entity, _handler)
  }
}
