// @sap/cds-runtime/lib/common/utils/templateProcessor.js
const _formatRowContext = ({ tKey, keyNames, row }) => {
  const keyValuePairs = keyNames.map(key => key + '=' + row[key])
  return tKey + '(' + keyValuePairs.join(',') + ')'
}

const _processElement = args => {
  const { processFn, row, key, elements, picked = {}, isRoot, complex = false, rowKeysGenerator, pathSegments } = args
  const element = elements[key]
  const { plain } = picked

  if (plain) {
    if (!complex && typeof rowKeysGenerator === 'function') {
      pathSegments.push(key)
    }
    // skip processing if row is undefined, it is not mandatory, and null is allowed
    // where should this go? `if (row && (row[key] !== undefined || element._isMandatory || element.notNull))`
    processFn(row, key, element, plain, isRoot, pathSegments)
  }
}

const _processRow = ({ processFn, row, template, tKey, tValue, isRoot, extraKeys, rowKeysGenerator, pathSegments }) => {
  const { template: subTemplate, picked } = tValue
  const key = tKey.split(' ').pop()

  const args = {
    processFn,
    row,
    key,
    elements: template.target.elements,
    picked,
    complex: !!subTemplate,
    isRoot,
    rowKeysGenerator,
    pathSegments,
    extraKeys
  }

  _processElement(args)

  // process deep
  if (subTemplate) {
    let subRows = row && row[key]
    subRows = Array.isArray(subRows) ? subRows : [subRows]
    const complexArgs = {
      processFn,
      rows: subRows,
      template: subTemplate,
      tKey: key,
      extraKeys,
      rowKeysGenerator,
      pathSegments
    }
    _processComplex(complexArgs)
  }
}

const _processComplex = ({ processFn, rows, template, tKey, extraKeys, rowKeysGenerator, pathSegments }) => {
  if (rows.length === 0) return
  let keyNames

  for (const row of rows) {
    if (row == null) continue

    let pathSegment
    let complexPathSegments

    if (typeof rowKeysGenerator === 'function') {
      keyNames = keyNames || (template.target.keys && Object.keys(template.target.keys)) || []
      rowKeysGenerator({ keyNames, row, template })
      const rowWithExtraKeys = { ...row, ...extraKeys }
      pathSegment = _formatRowContext({ tKey, keyNames, row: rowWithExtraKeys })
      complexPathSegments = [...pathSegments, pathSegment]
    }

    const args = {
      processFn,
      row,
      template,
      isRoot: false,
      extraKeys,
      rowKeysGenerator,
      pathSegments: complexPathSegments || pathSegments
    }

    templateProcessor(args)
  }
}

/**
 * @typedef {object} TemplateProcessorInfo
 * @property {entity} target
 * @property {Map} elements
 */

/**
 * @typedef {object} TemplateProcessor
 * @property {Function} processFn
 * @property {object} row
 * @property {TemplateProcessorInfo} template
 * @property {boolean} [isRoot=true]
 * @property {object} [extraKeys]
 * @property {Function} [rowKeysGenerator]
 * @property {string[]} [pathSegments=[]] - Path segments to relate the error message.
 * The path segments are used to build the error target (a relative resource path)
 */

/**
 * @param {TemplateProcessor} args
 */
const templateProcessor = ({
  processFn,
  row,
  template,
  isRoot = true,
  extraKeys,
  rowKeysGenerator,
  pathSegments = []
}) => {
  for (const [tKey, tValue] of template.elements) {
    const args = {
      processFn,
      row,
      template,
      tKey,
      tValue,
      isRoot,
      extraKeys,
      rowKeysGenerator,
      pathSegments: [...pathSegments]
    }

    _processRow(args)
  }
}

module.exports = templateProcessor
