// @sap/cds-runtime/lib/common/auth/strategies/xsuaa.js
const cds = require('../../../cds')

const _require = require('../../utils/require')
const uaaUtils = require('./utils/uaa')
const xssecUtils = require('./utils/xssec')

// use _require for a better error message
const { JWTStrategy: JS } = _require('@sap/xssec')

class XSUAAStrategy extends JS {
  constructor(uaa) {
    super(uaaUtils.getCredentials(uaa))
    this.name = 'xsuaa'
  }

  authenticate(req, options) {
    // monkey patch success
    const _success = this.success
    this.success = (user, info) => {
      // create cds.User
      user = new cds.User(
        Object.assign(
          { id: xssecUtils.getUserId(user, info) },
          {
            _roles: xssecUtils.getRoles(['any', 'identified-user'], info),
            attr: xssecUtils.getAttrForXSSEC(info),
            tenant: xssecUtils.getTenant(info) || null
          }
        )
      )

      // set _req for locale getter
      Object.defineProperty(user, '_req', { enumerable: false, value: req })

      // call "super.success"
      _success(user, info)
    }

    super.authenticate(req, options)
  }
}

module.exports = XSUAAStrategy
