// @sap/cds-runtime/lib/remote/Service.js
const cds = require('../cds')
const LOG = cds.log('remote')

// REVISIT: use cds.log's logger in cloud sdk

// disable sdk logger if not in debug mode
if (!LOG._debug) {
  const sdkUtils = require('@sap-cloud-sdk/util')
  sdkUtils.setGlobalLogLevel('error')
}

const { resolveView, getTransition } = require('../common/utils/resolveView')
const { getKind, run, getDestination, getAdditionalOptions, getReqOptions, postProcess } = require('./utils/client')

const _checkProduction = destination => {
  if (!destination && process.env.NODE_ENV === 'production') {
    throw new Error('In production mode it is required to set `options.destination`')
  }
}

const _setHeaders = (defaultHeaders, req) => {
  // REVISIT: If header "x-cds-incoming-protocol" is not present, custom headers were provided.
  if (req.headers['x-cds-incoming-protocol']) {
    return defaultHeaders
  }

  return Object.assign(
    defaultHeaders,
    Object.keys(req.headers).reduce((acc, cur) => {
      acc[cur.toLowerCase()] = req.headers[cur]
      return acc
    }, {})
  )
}

class RemoteService extends cds.Service {
  init() {
    this.destination = this.options.credentials && this.options.credentials.destination
    this.requestTimeout = this.options.credentials && this.options.credentials.requestTimeout
    if (this.requestTimeout === null || this.requestTimeout === undefined) this.requestTimeout = 60000
    this.path = this.options.credentials && this.options.credentials.path
    this.datasource = this.options.datasource
    this.kind = getKind(this.options) // TODO: Simplify

    _checkProduction(this.destination)

    this.on('*', async (req, next) => {
      let { query } = req
      if (!query && !(typeof req.path === 'string')) return next()
      if (typeof query === 'object' && this.model) query = resolveView(query, this.model, 'remote', true)
      const transitions =
        (typeof query === 'object' && (query.SELECT || query.INSERT || query.UPDATE || query.DELETE)._transitions) || []
      const resolvedTarget =
        (transitions.length && [transitions.length - 1].target) || getTransition(req.target, true).target
      if (!this.destination) this.destination = getDestination(this.model, this.datasource, this.options)
      const reqOptions = getReqOptions(req, query, this, resolvedTarget)
      reqOptions.headers = _setHeaders(reqOptions.headers, req)
      const additionalOptions = getAdditionalOptions(req, this.destination, this.kind, resolvedTarget)

      const result = await run(reqOptions, additionalOptions)
      return typeof query === 'object' ? postProcess(query, result) : result
    })
  }
}

module.exports = RemoteService
