// @sap/cds-runtime/lib/odata/odata2cqn.js
/*
 * Generated by PEG.js 0.10.0.
 *
 * http://pegjs.org/
 */

'use strict'

function peg$subclass(child, parent) {
  function ctor() {
    this.constructor = child
  }
  ctor.prototype = parent.prototype
  child.prototype = new ctor()
}

function peg$SyntaxError(message, expected, found, location) {
  this.message = message
  this.expected = expected
  this.found = found
  this.location = location
  this.name = 'SyntaxError'

  if (typeof Error.captureStackTrace === 'function') {
    Error.captureStackTrace(this, peg$SyntaxError)
  }
}

peg$subclass(peg$SyntaxError, Error)

peg$SyntaxError.buildMessage = function (expected, found) {
  var DESCRIBE_EXPECTATION_FNS = {
    literal: function (expectation) {
      return '"' + literalEscape(expectation.text) + '"'
    },

    class: function (expectation) {
      var escapedParts = '',
        i

      for (i = 0; i < expectation.parts.length; i++) {
        escapedParts +=
          expectation.parts[i] instanceof Array
            ? classEscape(expectation.parts[i][0]) + '-' + classEscape(expectation.parts[i][1])
            : classEscape(expectation.parts[i])
      }

      return '[' + (expectation.inverted ? '^' : '') + escapedParts + ']'
    },

    any: function (expectation) {
      return 'any character'
    },

    end: function (expectation) {
      return 'end of input'
    },

    other: function (expectation) {
      return expectation.description
    }
  }

  function hex(ch) {
    return ch.charCodeAt(0).toString(16).toUpperCase()
  }

  function literalEscape(s) {
    return s
      .replace(/\\/g, '\\\\')
      .replace(/"/g, '\\"')
      .replace(/\0/g, '\\0')
      .replace(/\t/g, '\\t')
      .replace(/\n/g, '\\n')
      .replace(/\r/g, '\\r')
      .replace(/[\x00-\x0F]/g, function (ch) {
        return '\\x0' + hex(ch)
      })
      .replace(/[\x10-\x1F\x7F-\x9F]/g, function (ch) {
        return '\\x' + hex(ch)
      })
  }

  function classEscape(s) {
    return s
      .replace(/\\/g, '\\\\')
      .replace(/\]/g, '\\]')
      .replace(/\^/g, '\\^')
      .replace(/-/g, '\\-')
      .replace(/\0/g, '\\0')
      .replace(/\t/g, '\\t')
      .replace(/\n/g, '\\n')
      .replace(/\r/g, '\\r')
      .replace(/[\x00-\x0F]/g, function (ch) {
        return '\\x0' + hex(ch)
      })
      .replace(/[\x10-\x1F\x7F-\x9F]/g, function (ch) {
        return '\\x' + hex(ch)
      })
  }

  function describeExpectation(expectation) {
    return DESCRIBE_EXPECTATION_FNS[expectation.type](expectation)
  }

  function describeExpected(expected) {
    var descriptions = new Array(expected.length),
      i,
      j

    for (i = 0; i < expected.length; i++) {
      descriptions[i] = describeExpectation(expected[i])
    }

    descriptions.sort()

    if (descriptions.length > 0) {
      for (i = 1, j = 1; i < descriptions.length; i++) {
        if (descriptions[i - 1] !== descriptions[i]) {
          descriptions[j] = descriptions[i]
          j++
        }
      }
      descriptions.length = j
    }

    switch (descriptions.length) {
      case 1:
        return descriptions[0]

      case 2:
        return descriptions[0] + ' or ' + descriptions[1]

      default:
        return descriptions.slice(0, -1).join(', ') + ', or ' + descriptions[descriptions.length - 1]
    }
  }

  function describeFound(found) {
    return found ? '"' + literalEscape(found) + '"' : 'end of input'
  }

  return 'Expected ' + describeExpected(expected) + ' but ' + describeFound(found) + ' found.'
}

function peg$parse(input, options) {
  options = options !== void 0 ? options : {}

  var peg$FAILED = {},
    peg$startRuleFunctions = { ODataRelativeURI: peg$parseODataRelativeURI },
    peg$startRuleFunction = peg$parseODataRelativeURI,
    peg$c0 = '/',
    peg$c1 = peg$literalExpectation('/', false),
    peg$c2 = function (p) {
      SELECT = { from: p }
    },
    peg$c3 = '?',
    peg$c4 = peg$literalExpectation('?', false),
    peg$c5 = '&',
    peg$c6 = peg$literalExpectation('&', false),
    peg$c7 = function () {
      if (SELECT.expand) {
        SELECT.columns = SELECT.expand
        delete SELECT.expand
      }
      // when $select and $expand was not defined
      if (!SELECT.columns) {
        SELECT.columns = [{ ref: '*' }]
      }
      // in case when have $expand but have not $select
      if (hasNotAny$select(SELECT.columns)) {
        SELECT.columns.unshift({ ref: '*' })
      }
      return { SELECT }
    },
    peg$c8 = '$count',
    peg$c9 = peg$literalExpectation('$count', false),
    peg$c10 = '$ref',
    peg$c11 = peg$literalExpectation('$ref', false),
    peg$c12 = '$value',
    peg$c13 = peg$literalExpectation('$value', false),
    peg$c14 = function (crv) {
      return { ref: [crv] }
    },
    peg$c15 = function (head, filter, p) {
      return p
    },
    peg$c16 = function (head, filter, tail) {
      const ref = [filter ? { id: head, where: filter[1] } : head]
      if (tail) {
        const more = tail.ref,
          [{ val }] = more
        if (val) ref[ref.length - 1] = { id: ref[ref.length - 1], where: [more.shift()] }
        ref.push(...more)
      }
      return { ref }
    },
    peg$c17 = function (val) {
      return [val]
    },
    peg$c18 = '=',
    peg$c19 = peg$literalExpectation('=', false),
    peg$c20 = function (ref, val, more) {
      const args = [ref, '=', val]
      if (more) args.push('and', ...more[1])
      return args
    },
    peg$c21 = '$select=',
    peg$c22 = peg$literalExpectation('$select=', false),
    peg$c23 = '$expand=',
    peg$c24 = peg$literalExpectation('$expand=', false),
    peg$c25 = '$filter=',
    peg$c26 = peg$literalExpectation('$filter=', false),
    peg$c27 = '$orderby=',
    peg$c28 = peg$literalExpectation('$orderby=', false),
    peg$c29 = '$top=',
    peg$c30 = peg$literalExpectation('$top=', false),
    peg$c31 = '$skip=',
    peg$c32 = peg$literalExpectation('$skip=', false),
    peg$c33 = '$search=',
    peg$c34 = peg$literalExpectation('$search=', false),
    peg$c35 = '$count=',
    peg$c36 = peg$literalExpectation('$count=', false),
    peg$c37 = '$apply=',
    peg$c38 = peg$literalExpectation('$apply=', false),
    peg$c39 = '*',
    peg$c40 = peg$literalExpectation('*', false),
    peg$c41 = function () {
      return { ref: '*' }
    },
    peg$c42 = function (col) {
      ;(SELECT.expand || (SELECT.expand = [])).push(col)
      return col
    },
    peg$c43 = function (c) {
      c.expand = ['*']
    },
    peg$c44 = function () {
      stack.push(SELECT)
      SELECT = SELECT.expand[SELECT.expand.length - 1]
      SELECT.expand = []
    },
    peg$c45 = ';',
    peg$c46 = peg$literalExpectation(';', false),
    peg$c47 = function () {
      SELECT = stack.pop()
    },
    peg$c48 = function (val) {
      ;(SELECT.limit || (SELECT.limit = {})).rows = { val }
    },
    peg$c49 = function (val) {
      ;(SELECT.limit || (SELECT.limit = {})).offset = { val }
    },
    peg$c50 = function (p) {
      SELECT.search = p
    },
    peg$c51 = function (n) {
      return n ? [n] : []
    },
    peg$c52 = function (p, xpr) {
      p.push({ xpr })
    },
    peg$c53 = function (p, val) {
      p.push({ ref: [val] })
    },
    peg$c54 = function (p, val) {
      p.push({ val })
    },
    peg$c55 = function (p, ao, more) {
      p.push(ao, ...more)
    },
    peg$c56 = function (p) {
      return p
    },
    peg$c57 = function (p) {
      SELECT.where = p
    },
    peg$c58 = function (p, comp) {
      p.push(...comp)
    },
    peg$c59 = function (p, func) {
      p.push(func)
    },
    peg$c60 = 'asc',
    peg$c61 = peg$literalExpectation('asc', false),
    peg$c62 = 'desc',
    peg$c63 = peg$literalExpectation('desc', false),
    peg$c64 = function (ref, s) {
      return s
    },
    peg$c65 = function (ref, sort) {
      const appendObj = $(ref, sort && { sort })
      SELECT.orderBy = SELECT.orderBy ? [...SELECT.orderBy, appendObj] : [appendObj]
    },
    peg$c66 = 'true',
    peg$c67 = peg$literalExpectation('true', false),
    peg$c68 = 'false',
    peg$c69 = peg$literalExpectation('false', false),
    peg$c70 = function (val) {
      if (val === 'true') {
        SELECT.count = true
      }
    },
    peg$c71 = /^[a-zA-Z]/,
    peg$c72 = peg$classExpectation(
      [
        ['a', 'z'],
        ['A', 'Z']
      ],
      false,
      false
    ),
    peg$c73 = /^[a-zA-Z0-9]/,
    peg$c74 = peg$classExpectation(
      [
        ['a', 'z'],
        ['A', 'Z'],
        ['0', '9']
      ],
      false,
      false
    ),
    peg$c75 = /^[^&]/,
    peg$c76 = peg$classExpectation(['&'], true, false),
    peg$c77 = peg$otherExpectation('a comparison'),
    peg$c78 = 'eq',
    peg$c79 = peg$literalExpectation('eq', false),
    peg$c80 = 'ne',
    peg$c81 = peg$literalExpectation('ne', false),
    peg$c82 = 'lt',
    peg$c83 = peg$literalExpectation('lt', false),
    peg$c84 = 'gt',
    peg$c85 = peg$literalExpectation('gt', false),
    peg$c86 = 'le',
    peg$c87 = peg$literalExpectation('le', false),
    peg$c88 = 'ge',
    peg$c89 = peg$literalExpectation('ge', false),
    peg$c90 = function (a, o, b) {
      const op = { eq: '=', ne: '!=', lt: '<', gt: '>', le: '<=', ge: '>=' }[o] || o
      return [a, op, b]
    },
    peg$c91 = 'add',
    peg$c92 = peg$literalExpectation('add', false),
    peg$c93 = 'sub',
    peg$c94 = peg$literalExpectation('sub', false),
    peg$c95 = 'mul',
    peg$c96 = peg$literalExpectation('mul', false),
    peg$c97 = 'div',
    peg$c98 = peg$literalExpectation('div', false),
    peg$c99 = 'mod',
    peg$c100 = peg$literalExpectation('mod', false),
    peg$c101 = peg$otherExpectation('an operand'),
    peg$c102 = peg$otherExpectation('a reference'),
    peg$c103 = function (head, n) {
      return n
    },
    peg$c104 = function (head, tail) {
      return { ref: [head, ...tail] }
    },
    peg$c105 = function (val) {
      return typeof val === 'number' ? { val } : { val, literal: 'number' }
    },
    peg$c106 = function (val) {
      return { val }
    },
    peg$c107 = peg$otherExpectation('a function call'),
    peg$c108 = /^[a-z]/,
    peg$c109 = peg$classExpectation([['a', 'z']], false, false),
    peg$c110 = function (func, a, o) {
      return o
    },
    peg$c111 = function (func, a, more) {
      if (!strict || func in strict.functions) return { func, args: [a, ...more] }
      else error("'" + func + "' is an unknown function in OData URL spec (strict mode)")
    },
    peg$c112 = peg$otherExpectation('a boolean function'),
    peg$c113 = 'contains',
    peg$c114 = peg$literalExpectation('contains', false),
    peg$c115 = 'endswith',
    peg$c116 = peg$literalExpectation('endswith', false),
    peg$c117 = 'startswith',
    peg$c118 = peg$literalExpectation('startswith', false),
    peg$c119 = function (func, a, b) {
      return { func, args: [a, b] }
    },
    peg$c120 = 'not',
    peg$c121 = peg$literalExpectation('not', true),
    peg$c122 = function () {
      return 'not'
    },
    peg$c123 = 'and',
    peg$c124 = peg$literalExpectation('and', true),
    peg$c125 = function () {
      return 'and'
    },
    peg$c126 = 'or',
    peg$c127 = peg$literalExpectation('or', true),
    peg$c128 = function () {
      return 'or'
    },
    peg$c129 = 'aggregate',
    peg$c130 = peg$literalExpectation('aggregate', false),
    peg$c131 = 'groupby',
    peg$c132 = peg$literalExpectation('groupby', false),
    peg$c133 = 'filter',
    peg$c134 = peg$literalExpectation('filter', false),
    peg$c135 = 'expand',
    peg$c136 = peg$literalExpectation('expand', false),
    peg$c137 = 'search',
    peg$c138 = peg$literalExpectation('search', false),
    peg$c139 = 'concat',
    peg$c140 = peg$literalExpectation('concat', false),
    peg$c141 = 'compute',
    peg$c142 = peg$literalExpectation('compute', false),
    peg$c143 = 'bottompercent',
    peg$c144 = peg$literalExpectation('bottompercent', false),
    peg$c145 = 'bottomsum',
    peg$c146 = peg$literalExpectation('bottomsum', false),
    peg$c147 = 'toppercent',
    peg$c148 = peg$literalExpectation('toppercent', false),
    peg$c149 = 'topsum',
    peg$c150 = peg$literalExpectation('topsum', false),
    peg$c151 = function (alias) {
      return { func: 'count', args: ['*'], as: alias }
    },
    peg$c152 = function (res) {
      ;(SELECT.expand || (SELECT.expand = [])).push(res)
    },
    peg$c153 = function (path, func, alias) {
      return { func, args: [path], as: alias }
    },
    peg$c154 = 'with',
    peg$c155 = peg$literalExpectation('with', false),
    peg$c156 = function (func) {
      return func
    },
    peg$c157 = 'from',
    peg$c158 = peg$literalExpectation('from', false),
    peg$c159 = 'as',
    peg$c160 = peg$literalExpectation('as', false),
    peg$c161 = function (alias) {
      return alias
    },
    peg$c162 = function (val) {
      ;(SELECT.groupBy || (SELECT.groupBy = [])).push(val)
    },
    peg$c163 = 'rollup',
    peg$c164 = peg$literalExpectation('rollup', false),
    peg$c165 = '$all',
    peg$c166 = peg$literalExpectation('$all', false),
    peg$c167 = 'topcount',
    peg$c168 = peg$literalExpectation('topcount', false),
    peg$c169 = 'bottomcount',
    peg$c170 = peg$literalExpectation('bottomcount', false),
    peg$c171 = function (trafo, val, ref) {
      const oredrObj = { ...ref, sort: trafo === 'topcount' ? 'desc' : 'asc' }
      SELECT.orderBy = SELECT.orderBy ? [...SELECT.orderBy, oredrObj] : [oredrObj]
      ;(SELECT.limit || (SELECT.limit = {})).rows = { val }
    },
    peg$c172 = 'identity',
    peg$c173 = peg$literalExpectation('identity', false),
    peg$c174 = peg$otherExpectation('Edm.String'),
    peg$c175 = "'",
    peg$c176 = peg$literalExpectation("'", false),
    peg$c177 = "''",
    peg$c178 = peg$literalExpectation("''", false),
    peg$c179 = /^[^']/,
    peg$c180 = peg$classExpectation(["'"], true, false),
    peg$c181 = function (s) {
      return s.replace(/''/g, "'")
    },
    peg$c182 = /^[+\-]/,
    peg$c183 = peg$classExpectation(['+', '-'], false, false),
    peg$c184 = /^[0-9]/,
    peg$c185 = peg$classExpectation([['0', '9']], false, false),
    peg$c186 = '.',
    peg$c187 = peg$literalExpectation('.', false),
    peg$c188 = 'e',
    peg$c189 = peg$literalExpectation('e', false),
    peg$c190 = function (s) {
      return safeNumber(s)
    },
    peg$c191 = function (s) {
      return parseInt(s)
    },
    peg$c192 = /^[_a-zA-Z0-9]/,
    peg$c193 = peg$classExpectation(['_', ['a', 'z'], ['A', 'Z'], ['0', '9']], false, false),
    peg$c194 = /^[0-9a-zA-Z]/,
    peg$c195 = peg$classExpectation(
      [
        ['0', '9'],
        ['a', 'z'],
        ['A', 'Z']
      ],
      false,
      false
    ),
    peg$c196 = '-',
    peg$c197 = peg$literalExpectation('-', false),
    peg$c198 = function (val) {
      return { val }
    },
    peg$c199 = ':',
    peg$c200 = peg$literalExpectation(':', false),
    peg$c201 = ',',
    peg$c202 = peg$literalExpectation(',', false),
    peg$c203 = '(',
    peg$c204 = peg$literalExpectation('(', false),
    peg$c205 = ')',
    peg$c206 = peg$literalExpectation(')', false),
    peg$c207 = peg$otherExpectation('optional whitespaces'),
    peg$c208 = /^[ \t\n]/,
    peg$c209 = peg$classExpectation([' ', '\t', '\n'], false, false),
    peg$c210 = peg$otherExpectation('mandatory whitespaces'),
    peg$currPos = 0,
    peg$savedPos = 0,
    peg$posDetailsCache = [{ line: 1, column: 1 }],
    peg$maxFailPos = 0,
    peg$maxFailExpected = [],
    peg$silentFails = 0,
    peg$result

  if ('startRule' in options) {
    if (!(options.startRule in peg$startRuleFunctions)) {
      throw new Error('Can\'t start parsing from rule "' + options.startRule + '".')
    }

    peg$startRuleFunction = peg$startRuleFunctions[options.startRule]
  }

  function text() {
    return input.substring(peg$savedPos, peg$currPos)
  }

  function location() {
    return peg$computeLocation(peg$savedPos, peg$currPos)
  }

  function expected(description, location) {
    location = location !== void 0 ? location : peg$computeLocation(peg$savedPos, peg$currPos)

    throw peg$buildStructuredError(
      [peg$otherExpectation(description)],
      input.substring(peg$savedPos, peg$currPos),
      location
    )
  }

  function error(message, location) {
    location = location !== void 0 ? location : peg$computeLocation(peg$savedPos, peg$currPos)

    throw peg$buildSimpleError(message, location)
  }

  function peg$literalExpectation(text, ignoreCase) {
    return { type: 'literal', text: text, ignoreCase: ignoreCase }
  }

  function peg$classExpectation(parts, inverted, ignoreCase) {
    return { type: 'class', parts: parts, inverted: inverted, ignoreCase: ignoreCase }
  }

  function peg$anyExpectation() {
    return { type: 'any' }
  }

  function peg$endExpectation() {
    return { type: 'end' }
  }

  function peg$otherExpectation(description) {
    return { type: 'other', description: description }
  }

  function peg$computePosDetails(pos) {
    var details = peg$posDetailsCache[pos],
      p

    if (details) {
      return details
    } else {
      p = pos - 1
      while (!peg$posDetailsCache[p]) {
        p--
      }

      details = peg$posDetailsCache[p]
      details = {
        line: details.line,
        column: details.column
      }

      while (p < pos) {
        if (input.charCodeAt(p) === 10) {
          details.line++
          details.column = 1
        } else {
          details.column++
        }

        p++
      }

      peg$posDetailsCache[pos] = details
      return details
    }
  }

  function peg$computeLocation(startPos, endPos) {
    var startPosDetails = peg$computePosDetails(startPos),
      endPosDetails = peg$computePosDetails(endPos)

    return {
      start: {
        offset: startPos,
        line: startPosDetails.line,
        column: startPosDetails.column
      },
      end: {
        offset: endPos,
        line: endPosDetails.line,
        column: endPosDetails.column
      }
    }
  }

  function peg$fail(expected) {
    if (peg$currPos < peg$maxFailPos) {
      return
    }

    if (peg$currPos > peg$maxFailPos) {
      peg$maxFailPos = peg$currPos
      peg$maxFailExpected = []
    }

    peg$maxFailExpected.push(expected)
  }

  function peg$buildSimpleError(message, location) {
    return new peg$SyntaxError(message, null, null, location)
  }

  function peg$buildStructuredError(expected, found, location) {
    return new peg$SyntaxError(peg$SyntaxError.buildMessage(expected, found), expected, found, location)
  }

  function peg$parseODataRelativeURI() {
    var s0, s1, s2, s3, s4, s5, s6, s7, s8, s9, s10, s11, s12, s13

    s0 = peg$currPos
    if (input.charCodeAt(peg$currPos) === 47) {
      s1 = peg$c0
      peg$currPos++
    } else {
      s1 = peg$FAILED
      if (peg$silentFails === 0) {
        peg$fail(peg$c1)
      }
    }
    if (s1 === peg$FAILED) {
      s1 = null
    }
    if (s1 !== peg$FAILED) {
      s2 = peg$currPos
      s3 = peg$parsepath()
      if (s3 !== peg$FAILED) {
        peg$savedPos = s2
        s3 = peg$c2(s3)
      }
      s2 = s3
      if (s2 !== peg$FAILED) {
        s3 = peg$currPos
        s4 = peg$parseo()
        if (s4 !== peg$FAILED) {
          if (input.charCodeAt(peg$currPos) === 63) {
            s5 = peg$c3
            peg$currPos++
          } else {
            s5 = peg$FAILED
            if (peg$silentFails === 0) {
              peg$fail(peg$c4)
            }
          }
          if (s5 !== peg$FAILED) {
            s6 = peg$parseo()
            if (s6 !== peg$FAILED) {
              s7 = peg$parseExpandOption()
              if (s7 !== peg$FAILED) {
                s8 = []
                s9 = peg$currPos
                s10 = peg$parseo()
                if (s10 !== peg$FAILED) {
                  if (input.charCodeAt(peg$currPos) === 38) {
                    s11 = peg$c5
                    peg$currPos++
                  } else {
                    s11 = peg$FAILED
                    if (peg$silentFails === 0) {
                      peg$fail(peg$c6)
                    }
                  }
                  if (s11 !== peg$FAILED) {
                    s12 = peg$parseo()
                    if (s12 !== peg$FAILED) {
                      s13 = peg$parseExpandOption()
                      if (s13 !== peg$FAILED) {
                        s10 = [s10, s11, s12, s13]
                        s9 = s10
                      } else {
                        peg$currPos = s9
                        s9 = peg$FAILED
                      }
                    } else {
                      peg$currPos = s9
                      s9 = peg$FAILED
                    }
                  } else {
                    peg$currPos = s9
                    s9 = peg$FAILED
                  }
                } else {
                  peg$currPos = s9
                  s9 = peg$FAILED
                }
                while (s9 !== peg$FAILED) {
                  s8.push(s9)
                  s9 = peg$currPos
                  s10 = peg$parseo()
                  if (s10 !== peg$FAILED) {
                    if (input.charCodeAt(peg$currPos) === 38) {
                      s11 = peg$c5
                      peg$currPos++
                    } else {
                      s11 = peg$FAILED
                      if (peg$silentFails === 0) {
                        peg$fail(peg$c6)
                      }
                    }
                    if (s11 !== peg$FAILED) {
                      s12 = peg$parseo()
                      if (s12 !== peg$FAILED) {
                        s13 = peg$parseExpandOption()
                        if (s13 !== peg$FAILED) {
                          s10 = [s10, s11, s12, s13]
                          s9 = s10
                        } else {
                          peg$currPos = s9
                          s9 = peg$FAILED
                        }
                      } else {
                        peg$currPos = s9
                        s9 = peg$FAILED
                      }
                    } else {
                      peg$currPos = s9
                      s9 = peg$FAILED
                    }
                  } else {
                    peg$currPos = s9
                    s9 = peg$FAILED
                  }
                }
                if (s8 !== peg$FAILED) {
                  s4 = [s4, s5, s6, s7, s8]
                  s3 = s4
                } else {
                  peg$currPos = s3
                  s3 = peg$FAILED
                }
              } else {
                peg$currPos = s3
                s3 = peg$FAILED
              }
            } else {
              peg$currPos = s3
              s3 = peg$FAILED
            }
          } else {
            peg$currPos = s3
            s3 = peg$FAILED
          }
        } else {
          peg$currPos = s3
          s3 = peg$FAILED
        }
        if (s3 === peg$FAILED) {
          s3 = null
        }
        if (s3 !== peg$FAILED) {
          s4 = peg$parseo()
          if (s4 !== peg$FAILED) {
            peg$savedPos = s0
            s1 = peg$c7()
            s0 = s1
          } else {
            peg$currPos = s0
            s0 = peg$FAILED
          }
        } else {
          peg$currPos = s0
          s0 = peg$FAILED
        }
      } else {
        peg$currPos = s0
        s0 = peg$FAILED
      }
    } else {
      peg$currPos = s0
      s0 = peg$FAILED
    }

    return s0
  }

  function peg$parsepath() {
    var s0, s1, s2, s3, s4, s5

    s0 = peg$currPos
    s1 = peg$currPos
    if (input.substr(peg$currPos, 6) === peg$c8) {
      s2 = peg$c8
      peg$currPos += 6
    } else {
      s2 = peg$FAILED
      if (peg$silentFails === 0) {
        peg$fail(peg$c9)
      }
    }
    if (s2 === peg$FAILED) {
      if (input.substr(peg$currPos, 4) === peg$c10) {
        s2 = peg$c10
        peg$currPos += 4
      } else {
        s2 = peg$FAILED
        if (peg$silentFails === 0) {
          peg$fail(peg$c11)
        }
      }
      if (s2 === peg$FAILED) {
        if (input.substr(peg$currPos, 6) === peg$c12) {
          s2 = peg$c12
          peg$currPos += 6
        } else {
          s2 = peg$FAILED
          if (peg$silentFails === 0) {
            peg$fail(peg$c13)
          }
        }
      }
    }
    if (s2 !== peg$FAILED) {
      s1 = input.substring(s1, peg$currPos)
    } else {
      s1 = s2
    }
    if (s1 !== peg$FAILED) {
      peg$savedPos = s0
      s1 = peg$c14(s1)
    }
    s0 = s1
    if (s0 === peg$FAILED) {
      s0 = peg$currPos
      s1 = peg$parseguid()
      if (s1 === peg$FAILED) {
        s1 = peg$parseidentifier()
        if (s1 === peg$FAILED) {
          s1 = peg$parseval()
        }
      }
      if (s1 !== peg$FAILED) {
        s2 = peg$currPos
        s3 = peg$parseOPEN()
        if (s3 !== peg$FAILED) {
          s4 = peg$parseargs()
          if (s4 !== peg$FAILED) {
            s5 = peg$parseCLOSE()
            if (s5 !== peg$FAILED) {
              s3 = [s3, s4, s5]
              s2 = s3
            } else {
              peg$currPos = s2
              s2 = peg$FAILED
            }
          } else {
            peg$currPos = s2
            s2 = peg$FAILED
          }
        } else {
          peg$currPos = s2
          s2 = peg$FAILED
        }
        if (s2 === peg$FAILED) {
          s2 = null
        }
        if (s2 !== peg$FAILED) {
          s3 = peg$currPos
          if (input.charCodeAt(peg$currPos) === 47) {
            s4 = peg$c0
            peg$currPos++
          } else {
            s4 = peg$FAILED
            if (peg$silentFails === 0) {
              peg$fail(peg$c1)
            }
          }
          if (s4 !== peg$FAILED) {
            s5 = peg$parsepath()
            if (s5 !== peg$FAILED) {
              peg$savedPos = s3
              s4 = peg$c15(s1, s2, s5)
              s3 = s4
            } else {
              peg$currPos = s3
              s3 = peg$FAILED
            }
          } else {
            peg$currPos = s3
            s3 = peg$FAILED
          }
          if (s3 === peg$FAILED) {
            s3 = null
          }
          if (s3 !== peg$FAILED) {
            peg$savedPos = s0
            s1 = peg$c16(s1, s2, s3)
            s0 = s1
          } else {
            peg$currPos = s0
            s0 = peg$FAILED
          }
        } else {
          peg$currPos = s0
          s0 = peg$FAILED
        }
      } else {
        peg$currPos = s0
        s0 = peg$FAILED
      }
    }

    return s0
  }

  function peg$parseargs() {
    var s0, s1, s2, s3, s4, s5, s6, s7, s8

    s0 = peg$currPos
    s1 = peg$parseguid()
    if (s1 === peg$FAILED) {
      s1 = peg$parseval()
    }
    if (s1 !== peg$FAILED) {
      peg$savedPos = s0
      s1 = peg$c17(s1)
    }
    s0 = s1
    if (s0 === peg$FAILED) {
      s0 = peg$currPos
      s1 = peg$parseref()
      if (s1 !== peg$FAILED) {
        s2 = peg$parseo()
        if (s2 !== peg$FAILED) {
          if (input.charCodeAt(peg$currPos) === 61) {
            s3 = peg$c18
            peg$currPos++
          } else {
            s3 = peg$FAILED
            if (peg$silentFails === 0) {
              peg$fail(peg$c19)
            }
          }
          if (s3 !== peg$FAILED) {
            s4 = peg$parseo()
            if (s4 !== peg$FAILED) {
              s5 = peg$parseguid()
              if (s5 === peg$FAILED) {
                s5 = peg$parseval()
              }
              if (s5 !== peg$FAILED) {
                s6 = peg$currPos
                s7 = peg$parseCOMMA()
                if (s7 !== peg$FAILED) {
                  s8 = peg$parseargs()
                  if (s8 !== peg$FAILED) {
                    s7 = [s7, s8]
                    s6 = s7
                  } else {
                    peg$currPos = s6
                    s6 = peg$FAILED
                  }
                } else {
                  peg$currPos = s6
                  s6 = peg$FAILED
                }
                if (s6 === peg$FAILED) {
                  s6 = null
                }
                if (s6 !== peg$FAILED) {
                  peg$savedPos = s0
                  s1 = peg$c20(s1, s5, s6)
                  s0 = s1
                } else {
                  peg$currPos = s0
                  s0 = peg$FAILED
                }
              } else {
                peg$currPos = s0
                s0 = peg$FAILED
              }
            } else {
              peg$currPos = s0
              s0 = peg$FAILED
            }
          } else {
            peg$currPos = s0
            s0 = peg$FAILED
          }
        } else {
          peg$currPos = s0
          s0 = peg$FAILED
        }
      } else {
        peg$currPos = s0
        s0 = peg$FAILED
      }
    }

    return s0
  }

  function peg$parseExpandOption() {
    var s0, s1, s2, s3, s4, s5, s6, s7

    s0 = peg$currPos
    if (input.substr(peg$currPos, 8) === peg$c21) {
      s1 = peg$c21
      peg$currPos += 8
    } else {
      s1 = peg$FAILED
      if (peg$silentFails === 0) {
        peg$fail(peg$c22)
      }
    }
    if (s1 !== peg$FAILED) {
      s2 = peg$parseo()
      if (s2 !== peg$FAILED) {
        s3 = peg$parseselect()
        if (s3 !== peg$FAILED) {
          s4 = []
          s5 = peg$currPos
          s6 = peg$parseCOMMA()
          if (s6 !== peg$FAILED) {
            s7 = peg$parseselect()
            if (s7 !== peg$FAILED) {
              s6 = [s6, s7]
              s5 = s6
            } else {
              peg$currPos = s5
              s5 = peg$FAILED
            }
          } else {
            peg$currPos = s5
            s5 = peg$FAILED
          }
          while (s5 !== peg$FAILED) {
            s4.push(s5)
            s5 = peg$currPos
            s6 = peg$parseCOMMA()
            if (s6 !== peg$FAILED) {
              s7 = peg$parseselect()
              if (s7 !== peg$FAILED) {
                s6 = [s6, s7]
                s5 = s6
              } else {
                peg$currPos = s5
                s5 = peg$FAILED
              }
            } else {
              peg$currPos = s5
              s5 = peg$FAILED
            }
          }
          if (s4 !== peg$FAILED) {
            s1 = [s1, s2, s3, s4]
            s0 = s1
          } else {
            peg$currPos = s0
            s0 = peg$FAILED
          }
        } else {
          peg$currPos = s0
          s0 = peg$FAILED
        }
      } else {
        peg$currPos = s0
        s0 = peg$FAILED
      }
    } else {
      peg$currPos = s0
      s0 = peg$FAILED
    }
    if (s0 === peg$FAILED) {
      s0 = peg$currPos
      if (input.substr(peg$currPos, 8) === peg$c23) {
        s1 = peg$c23
        peg$currPos += 8
      } else {
        s1 = peg$FAILED
        if (peg$silentFails === 0) {
          peg$fail(peg$c24)
        }
      }
      if (s1 !== peg$FAILED) {
        s2 = peg$parseo()
        if (s2 !== peg$FAILED) {
          s3 = peg$parseexpand()
          if (s3 !== peg$FAILED) {
            s4 = []
            s5 = peg$currPos
            s6 = peg$parseCOMMA()
            if (s6 !== peg$FAILED) {
              s7 = peg$parseexpand()
              if (s7 !== peg$FAILED) {
                s6 = [s6, s7]
                s5 = s6
              } else {
                peg$currPos = s5
                s5 = peg$FAILED
              }
            } else {
              peg$currPos = s5
              s5 = peg$FAILED
            }
            while (s5 !== peg$FAILED) {
              s4.push(s5)
              s5 = peg$currPos
              s6 = peg$parseCOMMA()
              if (s6 !== peg$FAILED) {
                s7 = peg$parseexpand()
                if (s7 !== peg$FAILED) {
                  s6 = [s6, s7]
                  s5 = s6
                } else {
                  peg$currPos = s5
                  s5 = peg$FAILED
                }
              } else {
                peg$currPos = s5
                s5 = peg$FAILED
              }
            }
            if (s4 !== peg$FAILED) {
              s1 = [s1, s2, s3, s4]
              s0 = s1
            } else {
              peg$currPos = s0
              s0 = peg$FAILED
            }
          } else {
            peg$currPos = s0
            s0 = peg$FAILED
          }
        } else {
          peg$currPos = s0
          s0 = peg$FAILED
        }
      } else {
        peg$currPos = s0
        s0 = peg$FAILED
      }
      if (s0 === peg$FAILED) {
        s0 = peg$currPos
        if (input.substr(peg$currPos, 8) === peg$c25) {
          s1 = peg$c25
          peg$currPos += 8
        } else {
          s1 = peg$FAILED
          if (peg$silentFails === 0) {
            peg$fail(peg$c26)
          }
        }
        if (s1 !== peg$FAILED) {
          s2 = peg$parseo()
          if (s2 !== peg$FAILED) {
            s3 = peg$parsefilter()
            if (s3 !== peg$FAILED) {
              s1 = [s1, s2, s3]
              s0 = s1
            } else {
              peg$currPos = s0
              s0 = peg$FAILED
            }
          } else {
            peg$currPos = s0
            s0 = peg$FAILED
          }
        } else {
          peg$currPos = s0
          s0 = peg$FAILED
        }
        if (s0 === peg$FAILED) {
          s0 = peg$currPos
          if (input.substr(peg$currPos, 9) === peg$c27) {
            s1 = peg$c27
            peg$currPos += 9
          } else {
            s1 = peg$FAILED
            if (peg$silentFails === 0) {
              peg$fail(peg$c28)
            }
          }
          if (s1 !== peg$FAILED) {
            s2 = peg$parseo()
            if (s2 !== peg$FAILED) {
              s3 = peg$parseorderby()
              if (s3 !== peg$FAILED) {
                s4 = []
                s5 = peg$currPos
                s6 = peg$parseCOMMA()
                if (s6 !== peg$FAILED) {
                  s7 = peg$parseorderby()
                  if (s7 !== peg$FAILED) {
                    s6 = [s6, s7]
                    s5 = s6
                  } else {
                    peg$currPos = s5
                    s5 = peg$FAILED
                  }
                } else {
                  peg$currPos = s5
                  s5 = peg$FAILED
                }
                while (s5 !== peg$FAILED) {
                  s4.push(s5)
                  s5 = peg$currPos
                  s6 = peg$parseCOMMA()
                  if (s6 !== peg$FAILED) {
                    s7 = peg$parseorderby()
                    if (s7 !== peg$FAILED) {
                      s6 = [s6, s7]
                      s5 = s6
                    } else {
                      peg$currPos = s5
                      s5 = peg$FAILED
                    }
                  } else {
                    peg$currPos = s5
                    s5 = peg$FAILED
                  }
                }
                if (s4 !== peg$FAILED) {
                  s1 = [s1, s2, s3, s4]
                  s0 = s1
                } else {
                  peg$currPos = s0
                  s0 = peg$FAILED
                }
              } else {
                peg$currPos = s0
                s0 = peg$FAILED
              }
            } else {
              peg$currPos = s0
              s0 = peg$FAILED
            }
          } else {
            peg$currPos = s0
            s0 = peg$FAILED
          }
          if (s0 === peg$FAILED) {
            s0 = peg$currPos
            if (input.substr(peg$currPos, 5) === peg$c29) {
              s1 = peg$c29
              peg$currPos += 5
            } else {
              s1 = peg$FAILED
              if (peg$silentFails === 0) {
                peg$fail(peg$c30)
              }
            }
            if (s1 !== peg$FAILED) {
              s2 = peg$parseo()
              if (s2 !== peg$FAILED) {
                s3 = peg$parsetop()
                if (s3 !== peg$FAILED) {
                  s1 = [s1, s2, s3]
                  s0 = s1
                } else {
                  peg$currPos = s0
                  s0 = peg$FAILED
                }
              } else {
                peg$currPos = s0
                s0 = peg$FAILED
              }
            } else {
              peg$currPos = s0
              s0 = peg$FAILED
            }
            if (s0 === peg$FAILED) {
              s0 = peg$currPos
              if (input.substr(peg$currPos, 6) === peg$c31) {
                s1 = peg$c31
                peg$currPos += 6
              } else {
                s1 = peg$FAILED
                if (peg$silentFails === 0) {
                  peg$fail(peg$c32)
                }
              }
              if (s1 !== peg$FAILED) {
                s2 = peg$parseo()
                if (s2 !== peg$FAILED) {
                  s3 = peg$parseskip()
                  if (s3 !== peg$FAILED) {
                    s1 = [s1, s2, s3]
                    s0 = s1
                  } else {
                    peg$currPos = s0
                    s0 = peg$FAILED
                  }
                } else {
                  peg$currPos = s0
                  s0 = peg$FAILED
                }
              } else {
                peg$currPos = s0
                s0 = peg$FAILED
              }
              if (s0 === peg$FAILED) {
                s0 = peg$currPos
                if (input.substr(peg$currPos, 8) === peg$c33) {
                  s1 = peg$c33
                  peg$currPos += 8
                } else {
                  s1 = peg$FAILED
                  if (peg$silentFails === 0) {
                    peg$fail(peg$c34)
                  }
                }
                if (s1 !== peg$FAILED) {
                  s2 = peg$parseo()
                  if (s2 !== peg$FAILED) {
                    s3 = peg$parsesearch()
                    if (s3 !== peg$FAILED) {
                      s1 = [s1, s2, s3]
                      s0 = s1
                    } else {
                      peg$currPos = s0
                      s0 = peg$FAILED
                    }
                  } else {
                    peg$currPos = s0
                    s0 = peg$FAILED
                  }
                } else {
                  peg$currPos = s0
                  s0 = peg$FAILED
                }
                if (s0 === peg$FAILED) {
                  s0 = peg$currPos
                  if (input.substr(peg$currPos, 7) === peg$c35) {
                    s1 = peg$c35
                    peg$currPos += 7
                  } else {
                    s1 = peg$FAILED
                    if (peg$silentFails === 0) {
                      peg$fail(peg$c36)
                    }
                  }
                  if (s1 !== peg$FAILED) {
                    s2 = peg$parseo()
                    if (s2 !== peg$FAILED) {
                      s3 = peg$parsecount()
                      if (s3 !== peg$FAILED) {
                        s1 = [s1, s2, s3]
                        s0 = s1
                      } else {
                        peg$currPos = s0
                        s0 = peg$FAILED
                      }
                    } else {
                      peg$currPos = s0
                      s0 = peg$FAILED
                    }
                  } else {
                    peg$currPos = s0
                    s0 = peg$FAILED
                  }
                  if (s0 === peg$FAILED) {
                    s0 = peg$currPos
                    if (input.substr(peg$currPos, 7) === peg$c37) {
                      s1 = peg$c37
                      peg$currPos += 7
                    } else {
                      s1 = peg$FAILED
                      if (peg$silentFails === 0) {
                        peg$fail(peg$c38)
                      }
                    }
                    if (s1 !== peg$FAILED) {
                      s2 = peg$parseo()
                      if (s2 !== peg$FAILED) {
                        s3 = peg$parseapply()
                        if (s3 !== peg$FAILED) {
                          s1 = [s1, s2, s3]
                          s0 = s1
                        } else {
                          peg$currPos = s0
                          s0 = peg$FAILED
                        }
                      } else {
                        peg$currPos = s0
                        s0 = peg$FAILED
                      }
                    } else {
                      peg$currPos = s0
                      s0 = peg$FAILED
                    }
                    if (s0 === peg$FAILED) {
                      s0 = peg$parsecustom()
                    }
                  }
                }
              }
            }
          }
        }
      }
    }

    return s0
  }

  function peg$parseselect() {
    var s0, s1, s2

    s0 = peg$currPos
    s1 = peg$parseref()
    if (s1 === peg$FAILED) {
      s1 = peg$currPos
      if (input.charCodeAt(peg$currPos) === 42) {
        s2 = peg$c39
        peg$currPos++
      } else {
        s2 = peg$FAILED
        if (peg$silentFails === 0) {
          peg$fail(peg$c40)
        }
      }
      if (s2 !== peg$FAILED) {
        peg$savedPos = s1
        s2 = peg$c41()
      }
      s1 = s2
    }
    if (s1 !== peg$FAILED) {
      peg$savedPos = s0
      s1 = peg$c42(s1)
    }
    s0 = s1

    return s0
  }

  function peg$parseexpand() {
    var s0, s1, s2, s3, s4, s5, s6, s7, s8, s9, s10

    s0 = peg$currPos
    s1 = peg$currPos
    s2 = peg$parseselect()
    if (s2 !== peg$FAILED) {
      peg$savedPos = s1
      s2 = peg$c43(s2)
    }
    s1 = s2
    if (s1 !== peg$FAILED) {
      s2 = peg$currPos
      s3 = peg$currPos
      s4 = peg$parseOPEN()
      if (s4 !== peg$FAILED) {
        peg$savedPos = s3
        s4 = peg$c44()
      }
      s3 = s4
      if (s3 !== peg$FAILED) {
        s4 = peg$parseExpandOption()
        if (s4 !== peg$FAILED) {
          s5 = []
          s6 = peg$currPos
          s7 = peg$parseo()
          if (s7 !== peg$FAILED) {
            if (input.charCodeAt(peg$currPos) === 59) {
              s8 = peg$c45
              peg$currPos++
            } else {
              s8 = peg$FAILED
              if (peg$silentFails === 0) {
                peg$fail(peg$c46)
              }
            }
            if (s8 !== peg$FAILED) {
              s9 = peg$parseo()
              if (s9 !== peg$FAILED) {
                s10 = peg$parseExpandOption()
                if (s10 !== peg$FAILED) {
                  s7 = [s7, s8, s9, s10]
                  s6 = s7
                } else {
                  peg$currPos = s6
                  s6 = peg$FAILED
                }
              } else {
                peg$currPos = s6
                s6 = peg$FAILED
              }
            } else {
              peg$currPos = s6
              s6 = peg$FAILED
            }
          } else {
            peg$currPos = s6
            s6 = peg$FAILED
          }
          while (s6 !== peg$FAILED) {
            s5.push(s6)
            s6 = peg$currPos
            s7 = peg$parseo()
            if (s7 !== peg$FAILED) {
              if (input.charCodeAt(peg$currPos) === 59) {
                s8 = peg$c45
                peg$currPos++
              } else {
                s8 = peg$FAILED
                if (peg$silentFails === 0) {
                  peg$fail(peg$c46)
                }
              }
              if (s8 !== peg$FAILED) {
                s9 = peg$parseo()
                if (s9 !== peg$FAILED) {
                  s10 = peg$parseExpandOption()
                  if (s10 !== peg$FAILED) {
                    s7 = [s7, s8, s9, s10]
                    s6 = s7
                  } else {
                    peg$currPos = s6
                    s6 = peg$FAILED
                  }
                } else {
                  peg$currPos = s6
                  s6 = peg$FAILED
                }
              } else {
                peg$currPos = s6
                s6 = peg$FAILED
              }
            } else {
              peg$currPos = s6
              s6 = peg$FAILED
            }
          }
          if (s5 !== peg$FAILED) {
            s6 = peg$currPos
            s7 = peg$parseCLOSE()
            if (s7 !== peg$FAILED) {
              peg$savedPos = s6
              s7 = peg$c47()
            }
            s6 = s7
            if (s6 !== peg$FAILED) {
              s3 = [s3, s4, s5, s6]
              s2 = s3
            } else {
              peg$currPos = s2
              s2 = peg$FAILED
            }
          } else {
            peg$currPos = s2
            s2 = peg$FAILED
          }
        } else {
          peg$currPos = s2
          s2 = peg$FAILED
        }
      } else {
        peg$currPos = s2
        s2 = peg$FAILED
      }
      if (s2 === peg$FAILED) {
        s2 = null
      }
      if (s2 !== peg$FAILED) {
        s3 = peg$currPos
        s4 = peg$parseCOMMA()
        if (s4 !== peg$FAILED) {
          s5 = peg$parseexpand()
          if (s5 !== peg$FAILED) {
            s4 = [s4, s5]
            s3 = s4
          } else {
            peg$currPos = s3
            s3 = peg$FAILED
          }
        } else {
          peg$currPos = s3
          s3 = peg$FAILED
        }
        if (s3 === peg$FAILED) {
          s3 = null
        }
        if (s3 !== peg$FAILED) {
          s1 = [s1, s2, s3]
          s0 = s1
        } else {
          peg$currPos = s0
          s0 = peg$FAILED
        }
      } else {
        peg$currPos = s0
        s0 = peg$FAILED
      }
    } else {
      peg$currPos = s0
      s0 = peg$FAILED
    }

    return s0
  }

  function peg$parsetop() {
    var s0, s1

    s0 = peg$currPos
    s1 = peg$parseinteger()
    if (s1 !== peg$FAILED) {
      peg$savedPos = s0
      s1 = peg$c48(s1)
    }
    s0 = s1

    return s0
  }

  function peg$parseskip() {
    var s0, s1

    s0 = peg$currPos
    s1 = peg$parseinteger()
    if (s1 !== peg$FAILED) {
      peg$savedPos = s0
      s1 = peg$c49(s1)
    }
    s0 = s1

    return s0
  }

  function peg$parsesearch() {
    var s0, s1

    s0 = peg$currPos
    s1 = peg$parsesearch_clause()
    if (s1 !== peg$FAILED) {
      peg$savedPos = s0
      s1 = peg$c50(s1)
    }
    s0 = s1

    return s0
  }

  function peg$parsesearch_clause() {
    var s0, s1, s2, s3, s4, s5, s6

    s0 = peg$currPos
    s1 = peg$currPos
    s2 = peg$parseNOT()
    if (s2 === peg$FAILED) {
      s2 = null
    }
    if (s2 !== peg$FAILED) {
      peg$savedPos = s1
      s2 = peg$c51(s2)
    }
    s1 = s2
    if (s1 !== peg$FAILED) {
      s2 = peg$currPos
      s3 = peg$parseOPEN()
      if (s3 !== peg$FAILED) {
        s4 = peg$parsesearch_clause()
        if (s4 !== peg$FAILED) {
          s5 = peg$parseCLOSE()
          if (s5 !== peg$FAILED) {
            peg$savedPos = s2
            s3 = peg$c52(s1, s4)
            s2 = s3
          } else {
            peg$currPos = s2
            s2 = peg$FAILED
          }
        } else {
          peg$currPos = s2
          s2 = peg$FAILED
        }
      } else {
        peg$currPos = s2
        s2 = peg$FAILED
      }
      if (s2 === peg$FAILED) {
        s2 = peg$currPos
        s3 = peg$parseidentifier()
        if (s3 !== peg$FAILED) {
          peg$savedPos = s2
          s3 = peg$c53(s1, s3)
        }
        s2 = s3
        if (s2 === peg$FAILED) {
          s2 = peg$currPos
          s3 = peg$parsestring()
          if (s3 !== peg$FAILED) {
            peg$savedPos = s2
            s3 = peg$c54(s1, s3)
          }
          s2 = s3
        }
      }
      if (s2 !== peg$FAILED) {
        s3 = []
        s4 = peg$currPos
        s5 = peg$parseAND()
        if (s5 === peg$FAILED) {
          s5 = peg$parseOR()
        }
        if (s5 !== peg$FAILED) {
          s6 = peg$parsesearch_clause()
          if (s6 !== peg$FAILED) {
            peg$savedPos = s4
            s5 = peg$c55(s1, s5, s6)
            s4 = s5
          } else {
            peg$currPos = s4
            s4 = peg$FAILED
          }
        } else {
          peg$currPos = s4
          s4 = peg$FAILED
        }
        while (s4 !== peg$FAILED) {
          s3.push(s4)
          s4 = peg$currPos
          s5 = peg$parseAND()
          if (s5 === peg$FAILED) {
            s5 = peg$parseOR()
          }
          if (s5 !== peg$FAILED) {
            s6 = peg$parsesearch_clause()
            if (s6 !== peg$FAILED) {
              peg$savedPos = s4
              s5 = peg$c55(s1, s5, s6)
              s4 = s5
            } else {
              peg$currPos = s4
              s4 = peg$FAILED
            }
          } else {
            peg$currPos = s4
            s4 = peg$FAILED
          }
        }
        if (s3 !== peg$FAILED) {
          peg$savedPos = s0
          s1 = peg$c56(s1)
          s0 = s1
        } else {
          peg$currPos = s0
          s0 = peg$FAILED
        }
      } else {
        peg$currPos = s0
        s0 = peg$FAILED
      }
    } else {
      peg$currPos = s0
      s0 = peg$FAILED
    }

    return s0
  }

  function peg$parsefilter() {
    var s0, s1

    s0 = peg$currPos
    s1 = peg$parsewhere_clause()
    if (s1 !== peg$FAILED) {
      peg$savedPos = s0
      s1 = peg$c57(s1)
    }
    s0 = s1

    return s0
  }

  function peg$parsewhere_clause() {
    var s0, s1, s2, s3, s4, s5, s6

    s0 = peg$currPos
    s1 = peg$currPos
    s2 = peg$parseNOT()
    if (s2 === peg$FAILED) {
      s2 = null
    }
    if (s2 !== peg$FAILED) {
      peg$savedPos = s1
      s2 = peg$c51(s2)
    }
    s1 = s2
    if (s1 !== peg$FAILED) {
      s2 = peg$currPos
      s3 = peg$parseOPEN()
      if (s3 !== peg$FAILED) {
        s4 = peg$parsewhere_clause()
        if (s4 !== peg$FAILED) {
          s5 = peg$parseCLOSE()
          if (s5 !== peg$FAILED) {
            peg$savedPos = s2
            s3 = peg$c52(s1, s4)
            s2 = s3
          } else {
            peg$currPos = s2
            s2 = peg$FAILED
          }
        } else {
          peg$currPos = s2
          s2 = peg$FAILED
        }
      } else {
        peg$currPos = s2
        s2 = peg$FAILED
      }
      if (s2 === peg$FAILED) {
        s2 = peg$currPos
        s3 = peg$parsecomparison()
        if (s3 !== peg$FAILED) {
          peg$savedPos = s2
          s3 = peg$c58(s1, s3)
        }
        s2 = s3
        if (s2 === peg$FAILED) {
          s2 = peg$currPos
          s3 = peg$parseboolish()
          if (s3 !== peg$FAILED) {
            peg$savedPos = s2
            s3 = peg$c59(s1, s3)
          }
          s2 = s3
        }
      }
      if (s2 !== peg$FAILED) {
        s3 = []
        s4 = peg$currPos
        s5 = peg$parseAND()
        if (s5 === peg$FAILED) {
          s5 = peg$parseOR()
        }
        if (s5 !== peg$FAILED) {
          s6 = peg$parsewhere_clause()
          if (s6 !== peg$FAILED) {
            peg$savedPos = s4
            s5 = peg$c55(s1, s5, s6)
            s4 = s5
          } else {
            peg$currPos = s4
            s4 = peg$FAILED
          }
        } else {
          peg$currPos = s4
          s4 = peg$FAILED
        }
        while (s4 !== peg$FAILED) {
          s3.push(s4)
          s4 = peg$currPos
          s5 = peg$parseAND()
          if (s5 === peg$FAILED) {
            s5 = peg$parseOR()
          }
          if (s5 !== peg$FAILED) {
            s6 = peg$parsewhere_clause()
            if (s6 !== peg$FAILED) {
              peg$savedPos = s4
              s5 = peg$c55(s1, s5, s6)
              s4 = s5
            } else {
              peg$currPos = s4
              s4 = peg$FAILED
            }
          } else {
            peg$currPos = s4
            s4 = peg$FAILED
          }
        }
        if (s3 !== peg$FAILED) {
          peg$savedPos = s0
          s1 = peg$c56(s1)
          s0 = s1
        } else {
          peg$currPos = s0
          s0 = peg$FAILED
        }
      } else {
        peg$currPos = s0
        s0 = peg$FAILED
      }
    } else {
      peg$currPos = s0
      s0 = peg$FAILED
    }

    return s0
  }

  function peg$parseorderby() {
    var s0, s1, s2, s3, s4, s5

    s0 = peg$currPos
    s1 = peg$parsefunction()
    if (s1 === peg$FAILED) {
      s1 = peg$parseref()
    }
    if (s1 !== peg$FAILED) {
      s2 = peg$currPos
      s3 = peg$parse_()
      if (s3 !== peg$FAILED) {
        s4 = peg$currPos
        if (input.substr(peg$currPos, 3) === peg$c60) {
          s5 = peg$c60
          peg$currPos += 3
        } else {
          s5 = peg$FAILED
          if (peg$silentFails === 0) {
            peg$fail(peg$c61)
          }
        }
        if (s5 === peg$FAILED) {
          if (input.substr(peg$currPos, 4) === peg$c62) {
            s5 = peg$c62
            peg$currPos += 4
          } else {
            s5 = peg$FAILED
            if (peg$silentFails === 0) {
              peg$fail(peg$c63)
            }
          }
        }
        if (s5 !== peg$FAILED) {
          s4 = input.substring(s4, peg$currPos)
        } else {
          s4 = s5
        }
        if (s4 !== peg$FAILED) {
          peg$savedPos = s2
          s3 = peg$c64(s1, s4)
          s2 = s3
        } else {
          peg$currPos = s2
          s2 = peg$FAILED
        }
      } else {
        peg$currPos = s2
        s2 = peg$FAILED
      }
      if (s2 === peg$FAILED) {
        s2 = null
      }
      if (s2 !== peg$FAILED) {
        peg$savedPos = s0
        s1 = peg$c65(s1, s2)
        s0 = s1
      } else {
        peg$currPos = s0
        s0 = peg$FAILED
      }
    } else {
      peg$currPos = s0
      s0 = peg$FAILED
    }

    return s0
  }

  function peg$parsecount() {
    var s0, s1

    s0 = peg$currPos
    if (input.substr(peg$currPos, 4) === peg$c66) {
      s1 = peg$c66
      peg$currPos += 4
    } else {
      s1 = peg$FAILED
      if (peg$silentFails === 0) {
        peg$fail(peg$c67)
      }
    }
    if (s1 === peg$FAILED) {
      if (input.substr(peg$currPos, 5) === peg$c68) {
        s1 = peg$c68
        peg$currPos += 5
      } else {
        s1 = peg$FAILED
        if (peg$silentFails === 0) {
          peg$fail(peg$c69)
        }
      }
    }
    if (s1 !== peg$FAILED) {
      peg$savedPos = s0
      s1 = peg$c70(s1)
    }
    s0 = s1

    return s0
  }

  function peg$parseapply() {
    var s0, s1, s2, s3, s4, s5

    s0 = peg$currPos
    s1 = peg$parseapplyTrafo()
    if (s1 !== peg$FAILED) {
      s2 = []
      s3 = peg$currPos
      if (input.charCodeAt(peg$currPos) === 47) {
        s4 = peg$c0
        peg$currPos++
      } else {
        s4 = peg$FAILED
        if (peg$silentFails === 0) {
          peg$fail(peg$c1)
        }
      }
      if (s4 !== peg$FAILED) {
        s5 = peg$parseapplyTrafo()
        if (s5 !== peg$FAILED) {
          s4 = [s4, s5]
          s3 = s4
        } else {
          peg$currPos = s3
          s3 = peg$FAILED
        }
      } else {
        peg$currPos = s3
        s3 = peg$FAILED
      }
      while (s3 !== peg$FAILED) {
        s2.push(s3)
        s3 = peg$currPos
        if (input.charCodeAt(peg$currPos) === 47) {
          s4 = peg$c0
          peg$currPos++
        } else {
          s4 = peg$FAILED
          if (peg$silentFails === 0) {
            peg$fail(peg$c1)
          }
        }
        if (s4 !== peg$FAILED) {
          s5 = peg$parseapplyTrafo()
          if (s5 !== peg$FAILED) {
            s4 = [s4, s5]
            s3 = s4
          } else {
            peg$currPos = s3
            s3 = peg$FAILED
          }
        } else {
          peg$currPos = s3
          s3 = peg$FAILED
        }
      }
      if (s2 !== peg$FAILED) {
        s1 = [s1, s2]
        s0 = s1
      } else {
        peg$currPos = s0
        s0 = peg$FAILED
      }
    } else {
      peg$currPos = s0
      s0 = peg$FAILED
    }

    return s0
  }

  function peg$parsecustom() {
    var s0, s1, s2, s3, s4, s5

    s0 = peg$currPos
    if (peg$c71.test(input.charAt(peg$currPos))) {
      s1 = input.charAt(peg$currPos)
      peg$currPos++
    } else {
      s1 = peg$FAILED
      if (peg$silentFails === 0) {
        peg$fail(peg$c72)
      }
    }
    if (s1 !== peg$FAILED) {
      s2 = []
      if (peg$c73.test(input.charAt(peg$currPos))) {
        s3 = input.charAt(peg$currPos)
        peg$currPos++
      } else {
        s3 = peg$FAILED
        if (peg$silentFails === 0) {
          peg$fail(peg$c74)
        }
      }
      while (s3 !== peg$FAILED) {
        s2.push(s3)
        if (peg$c73.test(input.charAt(peg$currPos))) {
          s3 = input.charAt(peg$currPos)
          peg$currPos++
        } else {
          s3 = peg$FAILED
          if (peg$silentFails === 0) {
            peg$fail(peg$c74)
          }
        }
      }
      if (s2 !== peg$FAILED) {
        if (input.charCodeAt(peg$currPos) === 61) {
          s3 = peg$c18
          peg$currPos++
        } else {
          s3 = peg$FAILED
          if (peg$silentFails === 0) {
            peg$fail(peg$c19)
          }
        }
        if (s3 !== peg$FAILED) {
          s4 = []
          if (peg$c75.test(input.charAt(peg$currPos))) {
            s5 = input.charAt(peg$currPos)
            peg$currPos++
          } else {
            s5 = peg$FAILED
            if (peg$silentFails === 0) {
              peg$fail(peg$c76)
            }
          }
          while (s5 !== peg$FAILED) {
            s4.push(s5)
            if (peg$c75.test(input.charAt(peg$currPos))) {
              s5 = input.charAt(peg$currPos)
              peg$currPos++
            } else {
              s5 = peg$FAILED
              if (peg$silentFails === 0) {
                peg$fail(peg$c76)
              }
            }
          }
          if (s4 !== peg$FAILED) {
            s1 = [s1, s2, s3, s4]
            s0 = s1
          } else {
            peg$currPos = s0
            s0 = peg$FAILED
          }
        } else {
          peg$currPos = s0
          s0 = peg$FAILED
        }
      } else {
        peg$currPos = s0
        s0 = peg$FAILED
      }
    } else {
      peg$currPos = s0
      s0 = peg$FAILED
    }

    return s0
  }

  function peg$parsecomparison() {
    var s0, s1, s2, s3, s4, s5

    peg$silentFails++
    s0 = peg$currPos
    s1 = peg$parseoperand()
    if (s1 !== peg$FAILED) {
      s2 = peg$parse_()
      if (s2 !== peg$FAILED) {
        s3 = peg$currPos
        if (input.substr(peg$currPos, 2) === peg$c78) {
          s4 = peg$c78
          peg$currPos += 2
        } else {
          s4 = peg$FAILED
          if (peg$silentFails === 0) {
            peg$fail(peg$c79)
          }
        }
        if (s4 === peg$FAILED) {
          if (input.substr(peg$currPos, 2) === peg$c80) {
            s4 = peg$c80
            peg$currPos += 2
          } else {
            s4 = peg$FAILED
            if (peg$silentFails === 0) {
              peg$fail(peg$c81)
            }
          }
          if (s4 === peg$FAILED) {
            if (input.substr(peg$currPos, 2) === peg$c82) {
              s4 = peg$c82
              peg$currPos += 2
            } else {
              s4 = peg$FAILED
              if (peg$silentFails === 0) {
                peg$fail(peg$c83)
              }
            }
            if (s4 === peg$FAILED) {
              if (input.substr(peg$currPos, 2) === peg$c84) {
                s4 = peg$c84
                peg$currPos += 2
              } else {
                s4 = peg$FAILED
                if (peg$silentFails === 0) {
                  peg$fail(peg$c85)
                }
              }
              if (s4 === peg$FAILED) {
                if (input.substr(peg$currPos, 2) === peg$c86) {
                  s4 = peg$c86
                  peg$currPos += 2
                } else {
                  s4 = peg$FAILED
                  if (peg$silentFails === 0) {
                    peg$fail(peg$c87)
                  }
                }
                if (s4 === peg$FAILED) {
                  if (input.substr(peg$currPos, 2) === peg$c88) {
                    s4 = peg$c88
                    peg$currPos += 2
                  } else {
                    s4 = peg$FAILED
                    if (peg$silentFails === 0) {
                      peg$fail(peg$c89)
                    }
                  }
                }
              }
            }
          }
        }
        if (s4 !== peg$FAILED) {
          s3 = input.substring(s3, peg$currPos)
        } else {
          s3 = s4
        }
        if (s3 !== peg$FAILED) {
          s4 = peg$parse_()
          if (s4 !== peg$FAILED) {
            s5 = peg$parseoperand()
            if (s5 !== peg$FAILED) {
              peg$savedPos = s0
              s1 = peg$c90(s1, s3, s5)
              s0 = s1
            } else {
              peg$currPos = s0
              s0 = peg$FAILED
            }
          } else {
            peg$currPos = s0
            s0 = peg$FAILED
          }
        } else {
          peg$currPos = s0
          s0 = peg$FAILED
        }
      } else {
        peg$currPos = s0
        s0 = peg$FAILED
      }
    } else {
      peg$currPos = s0
      s0 = peg$FAILED
    }
    peg$silentFails--
    if (s0 === peg$FAILED) {
      s1 = peg$FAILED
      if (peg$silentFails === 0) {
        peg$fail(peg$c77)
      }
    }

    return s0
  }

  function peg$parsemathCalc() {
    var s0, s1, s2, s3, s4, s5, s6, s7

    s0 = peg$currPos
    s1 = peg$parseoperand()
    if (s1 !== peg$FAILED) {
      s2 = []
      s3 = peg$currPos
      s4 = peg$parse_()
      if (s4 !== peg$FAILED) {
        if (input.substr(peg$currPos, 3) === peg$c91) {
          s5 = peg$c91
          peg$currPos += 3
        } else {
          s5 = peg$FAILED
          if (peg$silentFails === 0) {
            peg$fail(peg$c92)
          }
        }
        if (s5 === peg$FAILED) {
          if (input.substr(peg$currPos, 3) === peg$c93) {
            s5 = peg$c93
            peg$currPos += 3
          } else {
            s5 = peg$FAILED
            if (peg$silentFails === 0) {
              peg$fail(peg$c94)
            }
          }
          if (s5 === peg$FAILED) {
            if (input.substr(peg$currPos, 3) === peg$c95) {
              s5 = peg$c95
              peg$currPos += 3
            } else {
              s5 = peg$FAILED
              if (peg$silentFails === 0) {
                peg$fail(peg$c96)
              }
            }
            if (s5 === peg$FAILED) {
              if (input.substr(peg$currPos, 3) === peg$c97) {
                s5 = peg$c97
                peg$currPos += 3
              } else {
                s5 = peg$FAILED
                if (peg$silentFails === 0) {
                  peg$fail(peg$c98)
                }
              }
              if (s5 === peg$FAILED) {
                if (input.substr(peg$currPos, 3) === peg$c99) {
                  s5 = peg$c99
                  peg$currPos += 3
                } else {
                  s5 = peg$FAILED
                  if (peg$silentFails === 0) {
                    peg$fail(peg$c100)
                  }
                }
              }
            }
          }
        }
        if (s5 !== peg$FAILED) {
          s6 = peg$parse_()
          if (s6 !== peg$FAILED) {
            s7 = peg$parseoperand()
            if (s7 !== peg$FAILED) {
              s4 = [s4, s5, s6, s7]
              s3 = s4
            } else {
              peg$currPos = s3
              s3 = peg$FAILED
            }
          } else {
            peg$currPos = s3
            s3 = peg$FAILED
          }
        } else {
          peg$currPos = s3
          s3 = peg$FAILED
        }
      } else {
        peg$currPos = s3
        s3 = peg$FAILED
      }
      while (s3 !== peg$FAILED) {
        s2.push(s3)
        s3 = peg$currPos
        s4 = peg$parse_()
        if (s4 !== peg$FAILED) {
          if (input.substr(peg$currPos, 3) === peg$c91) {
            s5 = peg$c91
            peg$currPos += 3
          } else {
            s5 = peg$FAILED
            if (peg$silentFails === 0) {
              peg$fail(peg$c92)
            }
          }
          if (s5 === peg$FAILED) {
            if (input.substr(peg$currPos, 3) === peg$c93) {
              s5 = peg$c93
              peg$currPos += 3
            } else {
              s5 = peg$FAILED
              if (peg$silentFails === 0) {
                peg$fail(peg$c94)
              }
            }
            if (s5 === peg$FAILED) {
              if (input.substr(peg$currPos, 3) === peg$c95) {
                s5 = peg$c95
                peg$currPos += 3
              } else {
                s5 = peg$FAILED
                if (peg$silentFails === 0) {
                  peg$fail(peg$c96)
                }
              }
              if (s5 === peg$FAILED) {
                if (input.substr(peg$currPos, 3) === peg$c97) {
                  s5 = peg$c97
                  peg$currPos += 3
                } else {
                  s5 = peg$FAILED
                  if (peg$silentFails === 0) {
                    peg$fail(peg$c98)
                  }
                }
                if (s5 === peg$FAILED) {
                  if (input.substr(peg$currPos, 3) === peg$c99) {
                    s5 = peg$c99
                    peg$currPos += 3
                  } else {
                    s5 = peg$FAILED
                    if (peg$silentFails === 0) {
                      peg$fail(peg$c100)
                    }
                  }
                }
              }
            }
          }
          if (s5 !== peg$FAILED) {
            s6 = peg$parse_()
            if (s6 !== peg$FAILED) {
              s7 = peg$parseoperand()
              if (s7 !== peg$FAILED) {
                s4 = [s4, s5, s6, s7]
                s3 = s4
              } else {
                peg$currPos = s3
                s3 = peg$FAILED
              }
            } else {
              peg$currPos = s3
              s3 = peg$FAILED
            }
          } else {
            peg$currPos = s3
            s3 = peg$FAILED
          }
        } else {
          peg$currPos = s3
          s3 = peg$FAILED
        }
      }
      if (s2 !== peg$FAILED) {
        s1 = [s1, s2]
        s0 = s1
      } else {
        peg$currPos = s0
        s0 = peg$FAILED
      }
    } else {
      peg$currPos = s0
      s0 = peg$FAILED
    }

    return s0
  }

  function peg$parseoperand() {
    var s0, s1

    peg$silentFails++
    s0 = peg$parsefunction()
    if (s0 === peg$FAILED) {
      s0 = peg$parseref()
      if (s0 === peg$FAILED) {
        s0 = peg$parseval()
      }
    }
    peg$silentFails--
    if (s0 === peg$FAILED) {
      s1 = peg$FAILED
      if (peg$silentFails === 0) {
        peg$fail(peg$c101)
      }
    }

    return s0
  }

  function peg$parseref() {
    var s0, s1, s2, s3, s4, s5

    peg$silentFails++
    s0 = peg$currPos
    s1 = peg$parseidentifier()
    if (s1 !== peg$FAILED) {
      s2 = []
      s3 = peg$currPos
      if (input.charCodeAt(peg$currPos) === 47) {
        s4 = peg$c0
        peg$currPos++
      } else {
        s4 = peg$FAILED
        if (peg$silentFails === 0) {
          peg$fail(peg$c1)
        }
      }
      if (s4 !== peg$FAILED) {
        s5 = peg$parseidentifier()
        if (s5 !== peg$FAILED) {
          peg$savedPos = s3
          s4 = peg$c103(s1, s5)
          s3 = s4
        } else {
          peg$currPos = s3
          s3 = peg$FAILED
        }
      } else {
        peg$currPos = s3
        s3 = peg$FAILED
      }
      while (s3 !== peg$FAILED) {
        s2.push(s3)
        s3 = peg$currPos
        if (input.charCodeAt(peg$currPos) === 47) {
          s4 = peg$c0
          peg$currPos++
        } else {
          s4 = peg$FAILED
          if (peg$silentFails === 0) {
            peg$fail(peg$c1)
          }
        }
        if (s4 !== peg$FAILED) {
          s5 = peg$parseidentifier()
          if (s5 !== peg$FAILED) {
            peg$savedPos = s3
            s4 = peg$c103(s1, s5)
            s3 = s4
          } else {
            peg$currPos = s3
            s3 = peg$FAILED
          }
        } else {
          peg$currPos = s3
          s3 = peg$FAILED
        }
      }
      if (s2 !== peg$FAILED) {
        peg$savedPos = s0
        s1 = peg$c104(s1, s2)
        s0 = s1
      } else {
        peg$currPos = s0
        s0 = peg$FAILED
      }
    } else {
      peg$currPos = s0
      s0 = peg$FAILED
    }
    peg$silentFails--
    if (s0 === peg$FAILED) {
      s1 = peg$FAILED
      if (peg$silentFails === 0) {
        peg$fail(peg$c102)
      }
    }

    return s0
  }

  function peg$parseval() {
    var s0, s1

    s0 = peg$currPos
    s1 = peg$parsenumber()
    if (s1 !== peg$FAILED) {
      peg$savedPos = s0
      s1 = peg$c105(s1)
    }
    s0 = s1
    if (s0 === peg$FAILED) {
      s0 = peg$currPos
      s1 = peg$parsestring()
      if (s1 !== peg$FAILED) {
        peg$savedPos = s0
        s1 = peg$c106(s1)
      }
      s0 = s1
    }

    return s0
  }

  function peg$parsefunction() {
    var s0, s1, s2, s3, s4, s5, s6, s7

    peg$silentFails++
    s0 = peg$currPos
    s1 = peg$currPos
    s2 = []
    if (peg$c108.test(input.charAt(peg$currPos))) {
      s3 = input.charAt(peg$currPos)
      peg$currPos++
    } else {
      s3 = peg$FAILED
      if (peg$silentFails === 0) {
        peg$fail(peg$c109)
      }
    }
    if (s3 !== peg$FAILED) {
      while (s3 !== peg$FAILED) {
        s2.push(s3)
        if (peg$c108.test(input.charAt(peg$currPos))) {
          s3 = input.charAt(peg$currPos)
          peg$currPos++
        } else {
          s3 = peg$FAILED
          if (peg$silentFails === 0) {
            peg$fail(peg$c109)
          }
        }
      }
    } else {
      s2 = peg$FAILED
    }
    if (s2 !== peg$FAILED) {
      s1 = input.substring(s1, peg$currPos)
    } else {
      s1 = s2
    }
    if (s1 !== peg$FAILED) {
      s2 = peg$parseOPEN()
      if (s2 !== peg$FAILED) {
        s3 = peg$parseoperand()
        if (s3 !== peg$FAILED) {
          s4 = []
          s5 = peg$currPos
          s6 = peg$parseCOMMA()
          if (s6 !== peg$FAILED) {
            s7 = peg$parseoperand()
            if (s7 !== peg$FAILED) {
              peg$savedPos = s5
              s6 = peg$c110(s1, s3, s7)
              s5 = s6
            } else {
              peg$currPos = s5
              s5 = peg$FAILED
            }
          } else {
            peg$currPos = s5
            s5 = peg$FAILED
          }
          while (s5 !== peg$FAILED) {
            s4.push(s5)
            s5 = peg$currPos
            s6 = peg$parseCOMMA()
            if (s6 !== peg$FAILED) {
              s7 = peg$parseoperand()
              if (s7 !== peg$FAILED) {
                peg$savedPos = s5
                s6 = peg$c110(s1, s3, s7)
                s5 = s6
              } else {
                peg$currPos = s5
                s5 = peg$FAILED
              }
            } else {
              peg$currPos = s5
              s5 = peg$FAILED
            }
          }
          if (s4 !== peg$FAILED) {
            s5 = peg$parseCLOSE()
            if (s5 !== peg$FAILED) {
              peg$savedPos = s0
              s1 = peg$c111(s1, s3, s4)
              s0 = s1
            } else {
              peg$currPos = s0
              s0 = peg$FAILED
            }
          } else {
            peg$currPos = s0
            s0 = peg$FAILED
          }
        } else {
          peg$currPos = s0
          s0 = peg$FAILED
        }
      } else {
        peg$currPos = s0
        s0 = peg$FAILED
      }
    } else {
      peg$currPos = s0
      s0 = peg$FAILED
    }
    peg$silentFails--
    if (s0 === peg$FAILED) {
      s1 = peg$FAILED
      if (peg$silentFails === 0) {
        peg$fail(peg$c107)
      }
    }

    return s0
  }

  function peg$parseboolish() {
    var s0, s1, s2, s3, s4, s5, s6

    peg$silentFails++
    s0 = peg$currPos
    if (input.substr(peg$currPos, 8) === peg$c113) {
      s1 = peg$c113
      peg$currPos += 8
    } else {
      s1 = peg$FAILED
      if (peg$silentFails === 0) {
        peg$fail(peg$c114)
      }
    }
    if (s1 === peg$FAILED) {
      if (input.substr(peg$currPos, 8) === peg$c115) {
        s1 = peg$c115
        peg$currPos += 8
      } else {
        s1 = peg$FAILED
        if (peg$silentFails === 0) {
          peg$fail(peg$c116)
        }
      }
      if (s1 === peg$FAILED) {
        if (input.substr(peg$currPos, 10) === peg$c117) {
          s1 = peg$c117
          peg$currPos += 10
        } else {
          s1 = peg$FAILED
          if (peg$silentFails === 0) {
            peg$fail(peg$c118)
          }
        }
      }
    }
    if (s1 !== peg$FAILED) {
      s2 = peg$parseOPEN()
      if (s2 !== peg$FAILED) {
        s3 = peg$parseoperand()
        if (s3 !== peg$FAILED) {
          s4 = peg$parseCOMMA()
          if (s4 !== peg$FAILED) {
            s5 = peg$parseoperand()
            if (s5 !== peg$FAILED) {
              s6 = peg$parseCLOSE()
              if (s6 !== peg$FAILED) {
                peg$savedPos = s0
                s1 = peg$c119(s1, s3, s5)
                s0 = s1
              } else {
                peg$currPos = s0
                s0 = peg$FAILED
              }
            } else {
              peg$currPos = s0
              s0 = peg$FAILED
            }
          } else {
            peg$currPos = s0
            s0 = peg$FAILED
          }
        } else {
          peg$currPos = s0
          s0 = peg$FAILED
        }
      } else {
        peg$currPos = s0
        s0 = peg$FAILED
      }
    } else {
      peg$currPos = s0
      s0 = peg$FAILED
    }
    peg$silentFails--
    if (s0 === peg$FAILED) {
      s1 = peg$FAILED
      if (peg$silentFails === 0) {
        peg$fail(peg$c112)
      }
    }

    return s0
  }

  function peg$parseNOT() {
    var s0, s1, s2, s3

    s0 = peg$currPos
    s1 = peg$parseo()
    if (s1 !== peg$FAILED) {
      if (input.substr(peg$currPos, 3).toLowerCase() === peg$c120) {
        s2 = input.substr(peg$currPos, 3)
        peg$currPos += 3
      } else {
        s2 = peg$FAILED
        if (peg$silentFails === 0) {
          peg$fail(peg$c121)
        }
      }
      if (s2 !== peg$FAILED) {
        s3 = peg$parse_()
        if (s3 !== peg$FAILED) {
          peg$savedPos = s0
          s1 = peg$c122()
          s0 = s1
        } else {
          peg$currPos = s0
          s0 = peg$FAILED
        }
      } else {
        peg$currPos = s0
        s0 = peg$FAILED
      }
    } else {
      peg$currPos = s0
      s0 = peg$FAILED
    }

    return s0
  }

  function peg$parseAND() {
    var s0, s1, s2, s3

    s0 = peg$currPos
    s1 = peg$parse_()
    if (s1 !== peg$FAILED) {
      if (input.substr(peg$currPos, 3).toLowerCase() === peg$c123) {
        s2 = input.substr(peg$currPos, 3)
        peg$currPos += 3
      } else {
        s2 = peg$FAILED
        if (peg$silentFails === 0) {
          peg$fail(peg$c124)
        }
      }
      if (s2 !== peg$FAILED) {
        s3 = peg$parse_()
        if (s3 !== peg$FAILED) {
          peg$savedPos = s0
          s1 = peg$c125()
          s0 = s1
        } else {
          peg$currPos = s0
          s0 = peg$FAILED
        }
      } else {
        peg$currPos = s0
        s0 = peg$FAILED
      }
    } else {
      peg$currPos = s0
      s0 = peg$FAILED
    }

    return s0
  }

  function peg$parseOR() {
    var s0, s1, s2, s3

    s0 = peg$currPos
    s1 = peg$parse_()
    if (s1 !== peg$FAILED) {
      if (input.substr(peg$currPos, 2).toLowerCase() === peg$c126) {
        s2 = input.substr(peg$currPos, 2)
        peg$currPos += 2
      } else {
        s2 = peg$FAILED
        if (peg$silentFails === 0) {
          peg$fail(peg$c127)
        }
      }
      if (s2 !== peg$FAILED) {
        s3 = peg$parse_()
        if (s3 !== peg$FAILED) {
          peg$savedPos = s0
          s1 = peg$c128()
          s0 = s1
        } else {
          peg$currPos = s0
          s0 = peg$FAILED
        }
      } else {
        peg$currPos = s0
        s0 = peg$FAILED
      }
    } else {
      peg$currPos = s0
      s0 = peg$FAILED
    }

    return s0
  }

  function peg$parseapplyTrafo() {
    var s0, s1, s2

    s0 = peg$currPos
    if (input.substr(peg$currPos, 9) === peg$c129) {
      s1 = peg$c129
      peg$currPos += 9
    } else {
      s1 = peg$FAILED
      if (peg$silentFails === 0) {
        peg$fail(peg$c130)
      }
    }
    if (s1 !== peg$FAILED) {
      s2 = peg$parseaggregateTrafo()
      if (s2 !== peg$FAILED) {
        s1 = [s1, s2]
        s0 = s1
      } else {
        peg$currPos = s0
        s0 = peg$FAILED
      }
    } else {
      peg$currPos = s0
      s0 = peg$FAILED
    }
    if (s0 === peg$FAILED) {
      s0 = peg$currPos
      if (input.substr(peg$currPos, 7) === peg$c131) {
        s1 = peg$c131
        peg$currPos += 7
      } else {
        s1 = peg$FAILED
        if (peg$silentFails === 0) {
          peg$fail(peg$c132)
        }
      }
      if (s1 !== peg$FAILED) {
        s2 = peg$parsegroupbyTrafo()
        if (s2 !== peg$FAILED) {
          s1 = [s1, s2]
          s0 = s1
        } else {
          peg$currPos = s0
          s0 = peg$FAILED
        }
      } else {
        peg$currPos = s0
        s0 = peg$FAILED
      }
      if (s0 === peg$FAILED) {
        s0 = peg$currPos
        if (input.substr(peg$currPos, 6) === peg$c133) {
          s1 = peg$c133
          peg$currPos += 6
        } else {
          s1 = peg$FAILED
          if (peg$silentFails === 0) {
            peg$fail(peg$c134)
          }
        }
        if (s1 !== peg$FAILED) {
          s2 = peg$parsefilterTrafo()
          if (s2 !== peg$FAILED) {
            s1 = [s1, s2]
            s0 = s1
          } else {
            peg$currPos = s0
            s0 = peg$FAILED
          }
        } else {
          peg$currPos = s0
          s0 = peg$FAILED
        }
        if (s0 === peg$FAILED) {
          s0 = peg$parsecountTrafo()
          if (s0 === peg$FAILED) {
            s0 = peg$currPos
            if (input.substr(peg$currPos, 6) === peg$c135) {
              s1 = peg$c135
              peg$currPos += 6
            } else {
              s1 = peg$FAILED
              if (peg$silentFails === 0) {
                peg$fail(peg$c136)
              }
            }
            if (s1 !== peg$FAILED) {
              s2 = peg$parseexpandTrafo()
              if (s2 !== peg$FAILED) {
                s1 = [s1, s2]
                s0 = s1
              } else {
                peg$currPos = s0
                s0 = peg$FAILED
              }
            } else {
              peg$currPos = s0
              s0 = peg$FAILED
            }
            if (s0 === peg$FAILED) {
              s0 = peg$currPos
              if (input.substr(peg$currPos, 6) === peg$c137) {
                s1 = peg$c137
                peg$currPos += 6
              } else {
                s1 = peg$FAILED
                if (peg$silentFails === 0) {
                  peg$fail(peg$c138)
                }
              }
              if (s1 !== peg$FAILED) {
                s2 = peg$parsesearchTrafo()
                if (s2 !== peg$FAILED) {
                  s1 = [s1, s2]
                  s0 = s1
                } else {
                  peg$currPos = s0
                  s0 = peg$FAILED
                }
              } else {
                peg$currPos = s0
                s0 = peg$FAILED
              }
              if (s0 === peg$FAILED) {
                s0 = peg$currPos
                if (input.substr(peg$currPos, 6) === peg$c139) {
                  s1 = peg$c139
                  peg$currPos += 6
                } else {
                  s1 = peg$FAILED
                  if (peg$silentFails === 0) {
                    peg$fail(peg$c140)
                  }
                }
                if (s1 !== peg$FAILED) {
                  s2 = peg$parseconcatTrafo()
                  if (s2 !== peg$FAILED) {
                    s1 = [s1, s2]
                    s0 = s1
                  } else {
                    peg$currPos = s0
                    s0 = peg$FAILED
                  }
                } else {
                  peg$currPos = s0
                  s0 = peg$FAILED
                }
                if (s0 === peg$FAILED) {
                  s0 = peg$currPos
                  if (input.substr(peg$currPos, 7) === peg$c141) {
                    s1 = peg$c141
                    peg$currPos += 7
                  } else {
                    s1 = peg$FAILED
                    if (peg$silentFails === 0) {
                      peg$fail(peg$c142)
                    }
                  }
                  if (s1 !== peg$FAILED) {
                    s2 = peg$parsecomputeTrafo()
                    if (s2 !== peg$FAILED) {
                      s1 = [s1, s2]
                      s0 = s1
                    } else {
                      peg$currPos = s0
                      s0 = peg$FAILED
                    }
                  } else {
                    peg$currPos = s0
                    s0 = peg$FAILED
                  }
                  if (s0 === peg$FAILED) {
                    s0 = peg$currPos
                    if (input.substr(peg$currPos, 13) === peg$c143) {
                      s1 = peg$c143
                      peg$currPos += 13
                    } else {
                      s1 = peg$FAILED
                      if (peg$silentFails === 0) {
                        peg$fail(peg$c144)
                      }
                    }
                    if (s1 !== peg$FAILED) {
                      s2 = peg$parsecommonFuncTrafo()
                      if (s2 !== peg$FAILED) {
                        s1 = [s1, s2]
                        s0 = s1
                      } else {
                        peg$currPos = s0
                        s0 = peg$FAILED
                      }
                    } else {
                      peg$currPos = s0
                      s0 = peg$FAILED
                    }
                    if (s0 === peg$FAILED) {
                      s0 = peg$currPos
                      if (input.substr(peg$currPos, 9) === peg$c145) {
                        s1 = peg$c145
                        peg$currPos += 9
                      } else {
                        s1 = peg$FAILED
                        if (peg$silentFails === 0) {
                          peg$fail(peg$c146)
                        }
                      }
                      if (s1 !== peg$FAILED) {
                        s2 = peg$parsecommonFuncTrafo()
                        if (s2 !== peg$FAILED) {
                          s1 = [s1, s2]
                          s0 = s1
                        } else {
                          peg$currPos = s0
                          s0 = peg$FAILED
                        }
                      } else {
                        peg$currPos = s0
                        s0 = peg$FAILED
                      }
                      if (s0 === peg$FAILED) {
                        s0 = peg$currPos
                        if (input.substr(peg$currPos, 10) === peg$c147) {
                          s1 = peg$c147
                          peg$currPos += 10
                        } else {
                          s1 = peg$FAILED
                          if (peg$silentFails === 0) {
                            peg$fail(peg$c148)
                          }
                        }
                        if (s1 !== peg$FAILED) {
                          s2 = peg$parsecommonFuncTrafo()
                          if (s2 !== peg$FAILED) {
                            s1 = [s1, s2]
                            s0 = s1
                          } else {
                            peg$currPos = s0
                            s0 = peg$FAILED
                          }
                        } else {
                          peg$currPos = s0
                          s0 = peg$FAILED
                        }
                        if (s0 === peg$FAILED) {
                          s0 = peg$currPos
                          if (input.substr(peg$currPos, 6) === peg$c149) {
                            s1 = peg$c149
                            peg$currPos += 6
                          } else {
                            s1 = peg$FAILED
                            if (peg$silentFails === 0) {
                              peg$fail(peg$c150)
                            }
                          }
                          if (s1 !== peg$FAILED) {
                            s2 = peg$parsecommonFuncTrafo()
                            if (s2 !== peg$FAILED) {
                              s1 = [s1, s2]
                              s0 = s1
                            } else {
                              peg$currPos = s0
                              s0 = peg$FAILED
                            }
                          } else {
                            peg$currPos = s0
                            s0 = peg$FAILED
                          }
                          if (s0 === peg$FAILED) {
                            s0 = peg$parseidentityTrafo()
                          }
                        }
                      }
                    }
                  }
                }
              }
            }
          }
        }
      }
    }

    return s0
  }

  function peg$parseaggregateTrafo() {
    var s0, s1, s2, s3, s4, s5, s6, s7, s8, s9

    s0 = peg$currPos
    s1 = peg$parseOPEN()
    if (s1 !== peg$FAILED) {
      s2 = peg$parseo()
      if (s2 !== peg$FAILED) {
        s3 = peg$parseaggregateItem()
        if (s3 !== peg$FAILED) {
          s4 = []
          s5 = peg$currPos
          s6 = peg$parseo()
          if (s6 !== peg$FAILED) {
            s7 = peg$parseCOMMA()
            if (s7 !== peg$FAILED) {
              s8 = peg$parseo()
              if (s8 !== peg$FAILED) {
                s9 = peg$parseaggregateItem()
                if (s9 !== peg$FAILED) {
                  s6 = [s6, s7, s8, s9]
                  s5 = s6
                } else {
                  peg$currPos = s5
                  s5 = peg$FAILED
                }
              } else {
                peg$currPos = s5
                s5 = peg$FAILED
              }
            } else {
              peg$currPos = s5
              s5 = peg$FAILED
            }
          } else {
            peg$currPos = s5
            s5 = peg$FAILED
          }
          while (s5 !== peg$FAILED) {
            s4.push(s5)
            s5 = peg$currPos
            s6 = peg$parseo()
            if (s6 !== peg$FAILED) {
              s7 = peg$parseCOMMA()
              if (s7 !== peg$FAILED) {
                s8 = peg$parseo()
                if (s8 !== peg$FAILED) {
                  s9 = peg$parseaggregateItem()
                  if (s9 !== peg$FAILED) {
                    s6 = [s6, s7, s8, s9]
                    s5 = s6
                  } else {
                    peg$currPos = s5
                    s5 = peg$FAILED
                  }
                } else {
                  peg$currPos = s5
                  s5 = peg$FAILED
                }
              } else {
                peg$currPos = s5
                s5 = peg$FAILED
              }
            } else {
              peg$currPos = s5
              s5 = peg$FAILED
            }
          }
          if (s4 !== peg$FAILED) {
            s5 = peg$parseo()
            if (s5 !== peg$FAILED) {
              s6 = peg$parseCLOSE()
              if (s6 !== peg$FAILED) {
                s1 = [s1, s2, s3, s4, s5, s6]
                s0 = s1
              } else {
                peg$currPos = s0
                s0 = peg$FAILED
              }
            } else {
              peg$currPos = s0
              s0 = peg$FAILED
            }
          } else {
            peg$currPos = s0
            s0 = peg$FAILED
          }
        } else {
          peg$currPos = s0
          s0 = peg$FAILED
        }
      } else {
        peg$currPos = s0
        s0 = peg$FAILED
      }
    } else {
      peg$currPos = s0
      s0 = peg$FAILED
    }

    return s0
  }

  function peg$parseaggregateItem() {
    var s0, s1, s2, s3

    s0 = peg$currPos
    s1 = peg$currPos
    if (input.substr(peg$currPos, 6) === peg$c8) {
      s2 = peg$c8
      peg$currPos += 6
    } else {
      s2 = peg$FAILED
      if (peg$silentFails === 0) {
        peg$fail(peg$c9)
      }
    }
    if (s2 !== peg$FAILED) {
      s3 = peg$parseasAlias()
      if (s3 !== peg$FAILED) {
        peg$savedPos = s1
        s2 = peg$c151(s3)
        s1 = s2
      } else {
        peg$currPos = s1
        s1 = peg$FAILED
      }
    } else {
      peg$currPos = s1
      s1 = peg$FAILED
    }
    if (s1 === peg$FAILED) {
      s1 = peg$parseaggregateExpr()
    }
    if (s1 !== peg$FAILED) {
      peg$savedPos = s0
      s1 = peg$c152(s1)
    }
    s0 = s1

    return s0
  }

  function peg$parseaggregateExpr() {
    var s0, s1, s2, s3, s4

    s0 = peg$currPos
    s1 = peg$parseref()
    if (s1 !== peg$FAILED) {
      s2 = peg$parseaggregateWith()
      if (s2 !== peg$FAILED) {
        s3 = peg$parseaggregateFrom()
        if (s3 === peg$FAILED) {
          s3 = null
        }
        if (s3 !== peg$FAILED) {
          s4 = peg$parseasAlias()
          if (s4 !== peg$FAILED) {
            peg$savedPos = s0
            s1 = peg$c153(s1, s2, s4)
            s0 = s1
          } else {
            peg$currPos = s0
            s0 = peg$FAILED
          }
        } else {
          peg$currPos = s0
          s0 = peg$FAILED
        }
      } else {
        peg$currPos = s0
        s0 = peg$FAILED
      }
    } else {
      peg$currPos = s0
      s0 = peg$FAILED
    }
    if (s0 === peg$FAILED) {
      s0 = peg$currPos
      s1 = peg$parseidentifier()
      if (s1 !== peg$FAILED) {
        s2 = peg$parseOPEN()
        if (s2 !== peg$FAILED) {
          s3 = peg$parseaggregateExpr()
          if (s3 !== peg$FAILED) {
            s4 = peg$parseCLOSE()
            if (s4 !== peg$FAILED) {
              s1 = [s1, s2, s3, s4]
              s0 = s1
            } else {
              peg$currPos = s0
              s0 = peg$FAILED
            }
          } else {
            peg$currPos = s0
            s0 = peg$FAILED
          }
        } else {
          peg$currPos = s0
          s0 = peg$FAILED
        }
      } else {
        peg$currPos = s0
        s0 = peg$FAILED
      }
    }

    return s0
  }

  function peg$parseaggregateWith() {
    var s0, s1, s2, s3, s4, s5, s6

    s0 = peg$currPos
    s1 = peg$parse_()
    if (s1 !== peg$FAILED) {
      if (input.substr(peg$currPos, 4) === peg$c154) {
        s2 = peg$c154
        peg$currPos += 4
      } else {
        s2 = peg$FAILED
        if (peg$silentFails === 0) {
          peg$fail(peg$c155)
        }
      }
      if (s2 !== peg$FAILED) {
        s3 = peg$parse_()
        if (s3 !== peg$FAILED) {
          s4 = peg$currPos
          s5 = []
          if (peg$c108.test(input.charAt(peg$currPos))) {
            s6 = input.charAt(peg$currPos)
            peg$currPos++
          } else {
            s6 = peg$FAILED
            if (peg$silentFails === 0) {
              peg$fail(peg$c109)
            }
          }
          if (s6 !== peg$FAILED) {
            while (s6 !== peg$FAILED) {
              s5.push(s6)
              if (peg$c108.test(input.charAt(peg$currPos))) {
                s6 = input.charAt(peg$currPos)
                peg$currPos++
              } else {
                s6 = peg$FAILED
                if (peg$silentFails === 0) {
                  peg$fail(peg$c109)
                }
              }
            }
          } else {
            s5 = peg$FAILED
          }
          if (s5 !== peg$FAILED) {
            s4 = input.substring(s4, peg$currPos)
          } else {
            s4 = s5
          }
          if (s4 !== peg$FAILED) {
            peg$savedPos = s0
            s1 = peg$c156(s4)
            s0 = s1
          } else {
            peg$currPos = s0
            s0 = peg$FAILED
          }
        } else {
          peg$currPos = s0
          s0 = peg$FAILED
        }
      } else {
        peg$currPos = s0
        s0 = peg$FAILED
      }
    } else {
      peg$currPos = s0
      s0 = peg$FAILED
    }

    return s0
  }

  function peg$parseaggregateFrom() {
    var s0, s1, s2, s3, s4, s5, s6

    s0 = peg$currPos
    s1 = peg$parse_()
    if (s1 !== peg$FAILED) {
      if (input.substr(peg$currPos, 4) === peg$c157) {
        s2 = peg$c157
        peg$currPos += 4
      } else {
        s2 = peg$FAILED
        if (peg$silentFails === 0) {
          peg$fail(peg$c158)
        }
      }
      if (s2 !== peg$FAILED) {
        s3 = peg$parse_()
        if (s3 !== peg$FAILED) {
          s4 = peg$parseref()
          if (s4 !== peg$FAILED) {
            s5 = peg$parseaggregateWith()
            if (s5 !== peg$FAILED) {
              s6 = peg$parseaggregateFrom()
              if (s6 === peg$FAILED) {
                s6 = null
              }
              if (s6 !== peg$FAILED) {
                s1 = [s1, s2, s3, s4, s5, s6]
                s0 = s1
              } else {
                peg$currPos = s0
                s0 = peg$FAILED
              }
            } else {
              peg$currPos = s0
              s0 = peg$FAILED
            }
          } else {
            peg$currPos = s0
            s0 = peg$FAILED
          }
        } else {
          peg$currPos = s0
          s0 = peg$FAILED
        }
      } else {
        peg$currPos = s0
        s0 = peg$FAILED
      }
    } else {
      peg$currPos = s0
      s0 = peg$FAILED
    }

    return s0
  }

  function peg$parseasAlias() {
    var s0, s1, s2, s3, s4

    s0 = peg$currPos
    s1 = peg$parse_()
    if (s1 !== peg$FAILED) {
      if (input.substr(peg$currPos, 2) === peg$c159) {
        s2 = peg$c159
        peg$currPos += 2
      } else {
        s2 = peg$FAILED
        if (peg$silentFails === 0) {
          peg$fail(peg$c160)
        }
      }
      if (s2 !== peg$FAILED) {
        s3 = peg$parse_()
        if (s3 !== peg$FAILED) {
          s4 = peg$parseidentifier()
          if (s4 !== peg$FAILED) {
            peg$savedPos = s0
            s1 = peg$c161(s4)
            s0 = s1
          } else {
            peg$currPos = s0
            s0 = peg$FAILED
          }
        } else {
          peg$currPos = s0
          s0 = peg$FAILED
        }
      } else {
        peg$currPos = s0
        s0 = peg$FAILED
      }
    } else {
      peg$currPos = s0
      s0 = peg$FAILED
    }

    return s0
  }

  function peg$parsegroupbyTrafo() {
    var s0, s1, s2, s3, s4, s5, s6, s7, s8, s9, s10

    s0 = peg$currPos
    s1 = peg$parseOPEN()
    if (s1 !== peg$FAILED) {
      s2 = peg$parseo()
      if (s2 !== peg$FAILED) {
        s3 = peg$currPos
        s4 = peg$parseOPEN()
        if (s4 !== peg$FAILED) {
          s5 = peg$parsegroupByElem()
          if (s5 !== peg$FAILED) {
            s6 = []
            s7 = peg$currPos
            s8 = peg$parseCOMMA()
            if (s8 !== peg$FAILED) {
              s9 = peg$parseo()
              if (s9 !== peg$FAILED) {
                s10 = peg$parsegroupByElem()
                if (s10 !== peg$FAILED) {
                  s8 = [s8, s9, s10]
                  s7 = s8
                } else {
                  peg$currPos = s7
                  s7 = peg$FAILED
                }
              } else {
                peg$currPos = s7
                s7 = peg$FAILED
              }
            } else {
              peg$currPos = s7
              s7 = peg$FAILED
            }
            while (s7 !== peg$FAILED) {
              s6.push(s7)
              s7 = peg$currPos
              s8 = peg$parseCOMMA()
              if (s8 !== peg$FAILED) {
                s9 = peg$parseo()
                if (s9 !== peg$FAILED) {
                  s10 = peg$parsegroupByElem()
                  if (s10 !== peg$FAILED) {
                    s8 = [s8, s9, s10]
                    s7 = s8
                  } else {
                    peg$currPos = s7
                    s7 = peg$FAILED
                  }
                } else {
                  peg$currPos = s7
                  s7 = peg$FAILED
                }
              } else {
                peg$currPos = s7
                s7 = peg$FAILED
              }
            }
            if (s6 !== peg$FAILED) {
              s7 = peg$parseCLOSE()
              if (s7 !== peg$FAILED) {
                s4 = [s4, s5, s6, s7]
                s3 = s4
              } else {
                peg$currPos = s3
                s3 = peg$FAILED
              }
            } else {
              peg$currPos = s3
              s3 = peg$FAILED
            }
          } else {
            peg$currPos = s3
            s3 = peg$FAILED
          }
        } else {
          peg$currPos = s3
          s3 = peg$FAILED
        }
        if (s3 !== peg$FAILED) {
          s4 = peg$currPos
          s5 = peg$parseCOMMA()
          if (s5 !== peg$FAILED) {
            s6 = peg$parseo()
            if (s6 !== peg$FAILED) {
              s7 = peg$parseapply()
              if (s7 !== peg$FAILED) {
                s5 = [s5, s6, s7]
                s4 = s5
              } else {
                peg$currPos = s4
                s4 = peg$FAILED
              }
            } else {
              peg$currPos = s4
              s4 = peg$FAILED
            }
          } else {
            peg$currPos = s4
            s4 = peg$FAILED
          }
          if (s4 === peg$FAILED) {
            s4 = null
          }
          if (s4 !== peg$FAILED) {
            s5 = peg$parseo()
            if (s5 !== peg$FAILED) {
              s6 = peg$parseCLOSE()
              if (s6 !== peg$FAILED) {
                s1 = [s1, s2, s3, s4, s5, s6]
                s0 = s1
              } else {
                peg$currPos = s0
                s0 = peg$FAILED
              }
            } else {
              peg$currPos = s0
              s0 = peg$FAILED
            }
          } else {
            peg$currPos = s0
            s0 = peg$FAILED
          }
        } else {
          peg$currPos = s0
          s0 = peg$FAILED
        }
      } else {
        peg$currPos = s0
        s0 = peg$FAILED
      }
    } else {
      peg$currPos = s0
      s0 = peg$FAILED
    }

    return s0
  }

  function peg$parsegroupByElem() {
    var s0, s1

    s0 = peg$currPos
    s1 = peg$parserollupSpec()
    if (s1 === peg$FAILED) {
      s1 = peg$parseref()
    }
    if (s1 !== peg$FAILED) {
      peg$savedPos = s0
      s1 = peg$c162(s1)
    }
    s0 = s1

    return s0
  }

  function peg$parserollupSpec() {
    var s0, s1, s2, s3, s4, s5, s6, s7, s8, s9

    s0 = peg$currPos
    if (input.substr(peg$currPos, 6) === peg$c163) {
      s1 = peg$c163
      peg$currPos += 6
    } else {
      s1 = peg$FAILED
      if (peg$silentFails === 0) {
        peg$fail(peg$c164)
      }
    }
    if (s1 !== peg$FAILED) {
      s2 = peg$parseOPEN()
      if (s2 !== peg$FAILED) {
        s3 = peg$parseo()
        if (s3 !== peg$FAILED) {
          if (input.substr(peg$currPos, 4) === peg$c165) {
            s4 = peg$c165
            peg$currPos += 4
          } else {
            s4 = peg$FAILED
            if (peg$silentFails === 0) {
              peg$fail(peg$c166)
            }
          }
          if (s4 === peg$FAILED) {
            s4 = peg$parseref()
          }
          if (s4 !== peg$FAILED) {
            s5 = []
            s6 = peg$currPos
            s7 = peg$parseo()
            if (s7 !== peg$FAILED) {
              s8 = peg$parseCOMMA()
              if (s8 !== peg$FAILED) {
                s9 = peg$parseref()
                if (s9 !== peg$FAILED) {
                  s7 = [s7, s8, s9]
                  s6 = s7
                } else {
                  peg$currPos = s6
                  s6 = peg$FAILED
                }
              } else {
                peg$currPos = s6
                s6 = peg$FAILED
              }
            } else {
              peg$currPos = s6
              s6 = peg$FAILED
            }
            if (s6 !== peg$FAILED) {
              while (s6 !== peg$FAILED) {
                s5.push(s6)
                s6 = peg$currPos
                s7 = peg$parseo()
                if (s7 !== peg$FAILED) {
                  s8 = peg$parseCOMMA()
                  if (s8 !== peg$FAILED) {
                    s9 = peg$parseref()
                    if (s9 !== peg$FAILED) {
                      s7 = [s7, s8, s9]
                      s6 = s7
                    } else {
                      peg$currPos = s6
                      s6 = peg$FAILED
                    }
                  } else {
                    peg$currPos = s6
                    s6 = peg$FAILED
                  }
                } else {
                  peg$currPos = s6
                  s6 = peg$FAILED
                }
              }
            } else {
              s5 = peg$FAILED
            }
            if (s5 !== peg$FAILED) {
              s6 = peg$parseo()
              if (s6 !== peg$FAILED) {
                s7 = peg$parseCLOSE()
                if (s7 !== peg$FAILED) {
                  s1 = [s1, s2, s3, s4, s5, s6, s7]
                  s0 = s1
                } else {
                  peg$currPos = s0
                  s0 = peg$FAILED
                }
              } else {
                peg$currPos = s0
                s0 = peg$FAILED
              }
            } else {
              peg$currPos = s0
              s0 = peg$FAILED
            }
          } else {
            peg$currPos = s0
            s0 = peg$FAILED
          }
        } else {
          peg$currPos = s0
          s0 = peg$FAILED
        }
      } else {
        peg$currPos = s0
        s0 = peg$FAILED
      }
    } else {
      peg$currPos = s0
      s0 = peg$FAILED
    }

    return s0
  }

  function peg$parsefilterTrafo() {
    var s0, s1, s2, s3, s4, s5

    s0 = peg$currPos
    s1 = peg$parseOPEN()
    if (s1 !== peg$FAILED) {
      s2 = peg$parseo()
      if (s2 !== peg$FAILED) {
        s3 = peg$parsefilter()
        if (s3 !== peg$FAILED) {
          s4 = peg$parseo()
          if (s4 !== peg$FAILED) {
            s5 = peg$parseCLOSE()
            if (s5 !== peg$FAILED) {
              s1 = [s1, s2, s3, s4, s5]
              s0 = s1
            } else {
              peg$currPos = s0
              s0 = peg$FAILED
            }
          } else {
            peg$currPos = s0
            s0 = peg$FAILED
          }
        } else {
          peg$currPos = s0
          s0 = peg$FAILED
        }
      } else {
        peg$currPos = s0
        s0 = peg$FAILED
      }
    } else {
      peg$currPos = s0
      s0 = peg$FAILED
    }

    return s0
  }

  function peg$parsecountTrafo() {
    var s0, s1, s2, s3, s4, s5, s6, s7, s8, s9, s10

    s0 = peg$currPos
    if (input.substr(peg$currPos, 8) === peg$c167) {
      s1 = peg$c167
      peg$currPos += 8
    } else {
      s1 = peg$FAILED
      if (peg$silentFails === 0) {
        peg$fail(peg$c168)
      }
    }
    if (s1 === peg$FAILED) {
      if (input.substr(peg$currPos, 11) === peg$c169) {
        s1 = peg$c169
        peg$currPos += 11
      } else {
        s1 = peg$FAILED
        if (peg$silentFails === 0) {
          peg$fail(peg$c170)
        }
      }
    }
    if (s1 !== peg$FAILED) {
      s2 = peg$parseOPEN()
      if (s2 !== peg$FAILED) {
        s3 = peg$parseo()
        if (s3 !== peg$FAILED) {
          s4 = peg$parsenumber()
          if (s4 !== peg$FAILED) {
            s5 = peg$parseo()
            if (s5 !== peg$FAILED) {
              s6 = peg$parseCOMMA()
              if (s6 !== peg$FAILED) {
                s7 = peg$parseo()
                if (s7 !== peg$FAILED) {
                  s8 = peg$parseref()
                  if (s8 !== peg$FAILED) {
                    s9 = peg$parseo()
                    if (s9 !== peg$FAILED) {
                      s10 = peg$parseCLOSE()
                      if (s10 !== peg$FAILED) {
                        peg$savedPos = s0
                        s1 = peg$c171(s1, s4, s8)
                        s0 = s1
                      } else {
                        peg$currPos = s0
                        s0 = peg$FAILED
                      }
                    } else {
                      peg$currPos = s0
                      s0 = peg$FAILED
                    }
                  } else {
                    peg$currPos = s0
                    s0 = peg$FAILED
                  }
                } else {
                  peg$currPos = s0
                  s0 = peg$FAILED
                }
              } else {
                peg$currPos = s0
                s0 = peg$FAILED
              }
            } else {
              peg$currPos = s0
              s0 = peg$FAILED
            }
          } else {
            peg$currPos = s0
            s0 = peg$FAILED
          }
        } else {
          peg$currPos = s0
          s0 = peg$FAILED
        }
      } else {
        peg$currPos = s0
        s0 = peg$FAILED
      }
    } else {
      peg$currPos = s0
      s0 = peg$FAILED
    }

    return s0
  }

  function peg$parseexpandTrafo() {
    var s0, s1, s2, s3, s4, s5, s6, s7, s8, s9, s10, s11, s12, s13

    s0 = peg$currPos
    s1 = peg$parseOPEN()
    if (s1 !== peg$FAILED) {
      s2 = peg$parseo()
      if (s2 !== peg$FAILED) {
        s3 = peg$parseref()
        if (s3 !== peg$FAILED) {
          s4 = peg$parseo()
          if (s4 !== peg$FAILED) {
            s5 = peg$parseCOMMA()
            if (s5 !== peg$FAILED) {
              s6 = peg$parseo()
              if (s6 !== peg$FAILED) {
                s7 = peg$currPos
                s8 = peg$parseexpandTrafo()
                if (s8 !== peg$FAILED) {
                  s9 = []
                  s10 = peg$currPos
                  s11 = peg$parseo()
                  if (s11 !== peg$FAILED) {
                    s12 = peg$parseCOMMA()
                    if (s12 !== peg$FAILED) {
                      s13 = peg$parseexpandTrafo()
                      if (s13 !== peg$FAILED) {
                        s11 = [s11, s12, s13]
                        s10 = s11
                      } else {
                        peg$currPos = s10
                        s10 = peg$FAILED
                      }
                    } else {
                      peg$currPos = s10
                      s10 = peg$FAILED
                    }
                  } else {
                    peg$currPos = s10
                    s10 = peg$FAILED
                  }
                  while (s10 !== peg$FAILED) {
                    s9.push(s10)
                    s10 = peg$currPos
                    s11 = peg$parseo()
                    if (s11 !== peg$FAILED) {
                      s12 = peg$parseCOMMA()
                      if (s12 !== peg$FAILED) {
                        s13 = peg$parseexpandTrafo()
                        if (s13 !== peg$FAILED) {
                          s11 = [s11, s12, s13]
                          s10 = s11
                        } else {
                          peg$currPos = s10
                          s10 = peg$FAILED
                        }
                      } else {
                        peg$currPos = s10
                        s10 = peg$FAILED
                      }
                    } else {
                      peg$currPos = s10
                      s10 = peg$FAILED
                    }
                  }
                  if (s9 !== peg$FAILED) {
                    s8 = [s8, s9]
                    s7 = s8
                  } else {
                    peg$currPos = s7
                    s7 = peg$FAILED
                  }
                } else {
                  peg$currPos = s7
                  s7 = peg$FAILED
                }
                if (s7 === peg$FAILED) {
                  s7 = peg$currPos
                  s8 = peg$parsefilterTrafo()
                  if (s8 !== peg$FAILED) {
                    s9 = []
                    s10 = peg$currPos
                    s11 = peg$parseo()
                    if (s11 !== peg$FAILED) {
                      s12 = peg$parseCOMMA()
                      if (s12 !== peg$FAILED) {
                        s13 = peg$parseexpandTrafo()
                        if (s13 !== peg$FAILED) {
                          s11 = [s11, s12, s13]
                          s10 = s11
                        } else {
                          peg$currPos = s10
                          s10 = peg$FAILED
                        }
                      } else {
                        peg$currPos = s10
                        s10 = peg$FAILED
                      }
                    } else {
                      peg$currPos = s10
                      s10 = peg$FAILED
                    }
                    while (s10 !== peg$FAILED) {
                      s9.push(s10)
                      s10 = peg$currPos
                      s11 = peg$parseo()
                      if (s11 !== peg$FAILED) {
                        s12 = peg$parseCOMMA()
                        if (s12 !== peg$FAILED) {
                          s13 = peg$parseexpandTrafo()
                          if (s13 !== peg$FAILED) {
                            s11 = [s11, s12, s13]
                            s10 = s11
                          } else {
                            peg$currPos = s10
                            s10 = peg$FAILED
                          }
                        } else {
                          peg$currPos = s10
                          s10 = peg$FAILED
                        }
                      } else {
                        peg$currPos = s10
                        s10 = peg$FAILED
                      }
                    }
                    if (s9 !== peg$FAILED) {
                      s8 = [s8, s9]
                      s7 = s8
                    } else {
                      peg$currPos = s7
                      s7 = peg$FAILED
                    }
                  } else {
                    peg$currPos = s7
                    s7 = peg$FAILED
                  }
                }
                if (s7 !== peg$FAILED) {
                  s8 = peg$parseo()
                  if (s8 !== peg$FAILED) {
                    s9 = peg$parseCLOSE()
                    if (s9 !== peg$FAILED) {
                      s1 = [s1, s2, s3, s4, s5, s6, s7, s8, s9]
                      s0 = s1
                    } else {
                      peg$currPos = s0
                      s0 = peg$FAILED
                    }
                  } else {
                    peg$currPos = s0
                    s0 = peg$FAILED
                  }
                } else {
                  peg$currPos = s0
                  s0 = peg$FAILED
                }
              } else {
                peg$currPos = s0
                s0 = peg$FAILED
              }
            } else {
              peg$currPos = s0
              s0 = peg$FAILED
            }
          } else {
            peg$currPos = s0
            s0 = peg$FAILED
          }
        } else {
          peg$currPos = s0
          s0 = peg$FAILED
        }
      } else {
        peg$currPos = s0
        s0 = peg$FAILED
      }
    } else {
      peg$currPos = s0
      s0 = peg$FAILED
    }

    return s0
  }

  function peg$parsesearchTrafo() {
    var s0, s1, s2, s3, s4, s5

    s0 = peg$currPos
    s1 = peg$parseOPEN()
    if (s1 !== peg$FAILED) {
      s2 = peg$parseo()
      if (s2 !== peg$FAILED) {
        s3 = peg$parsesearch()
        if (s3 !== peg$FAILED) {
          s4 = peg$parseo()
          if (s4 !== peg$FAILED) {
            s5 = peg$parseCLOSE()
            if (s5 !== peg$FAILED) {
              s1 = [s1, s2, s3, s4, s5]
              s0 = s1
            } else {
              peg$currPos = s0
              s0 = peg$FAILED
            }
          } else {
            peg$currPos = s0
            s0 = peg$FAILED
          }
        } else {
          peg$currPos = s0
          s0 = peg$FAILED
        }
      } else {
        peg$currPos = s0
        s0 = peg$FAILED
      }
    } else {
      peg$currPos = s0
      s0 = peg$FAILED
    }

    return s0
  }

  function peg$parseconcatTrafo() {
    var s0, s1, s2, s3, s4, s5, s6, s7, s8, s9

    s0 = peg$currPos
    s1 = peg$parseOPEN()
    if (s1 !== peg$FAILED) {
      s2 = peg$parseo()
      if (s2 !== peg$FAILED) {
        s3 = peg$parseapply()
        if (s3 !== peg$FAILED) {
          s4 = []
          s5 = peg$currPos
          s6 = peg$parseo()
          if (s6 !== peg$FAILED) {
            s7 = peg$parseCOMMA()
            if (s7 !== peg$FAILED) {
              s8 = peg$parseo()
              if (s8 !== peg$FAILED) {
                s9 = peg$parseapply()
                if (s9 !== peg$FAILED) {
                  s6 = [s6, s7, s8, s9]
                  s5 = s6
                } else {
                  peg$currPos = s5
                  s5 = peg$FAILED
                }
              } else {
                peg$currPos = s5
                s5 = peg$FAILED
              }
            } else {
              peg$currPos = s5
              s5 = peg$FAILED
            }
          } else {
            peg$currPos = s5
            s5 = peg$FAILED
          }
          if (s5 !== peg$FAILED) {
            while (s5 !== peg$FAILED) {
              s4.push(s5)
              s5 = peg$currPos
              s6 = peg$parseo()
              if (s6 !== peg$FAILED) {
                s7 = peg$parseCOMMA()
                if (s7 !== peg$FAILED) {
                  s8 = peg$parseo()
                  if (s8 !== peg$FAILED) {
                    s9 = peg$parseapply()
                    if (s9 !== peg$FAILED) {
                      s6 = [s6, s7, s8, s9]
                      s5 = s6
                    } else {
                      peg$currPos = s5
                      s5 = peg$FAILED
                    }
                  } else {
                    peg$currPos = s5
                    s5 = peg$FAILED
                  }
                } else {
                  peg$currPos = s5
                  s5 = peg$FAILED
                }
              } else {
                peg$currPos = s5
                s5 = peg$FAILED
              }
            }
          } else {
            s4 = peg$FAILED
          }
          if (s4 !== peg$FAILED) {
            s5 = peg$parseo()
            if (s5 !== peg$FAILED) {
              s6 = peg$parseCLOSE()
              if (s6 !== peg$FAILED) {
                s1 = [s1, s2, s3, s4, s5, s6]
                s0 = s1
              } else {
                peg$currPos = s0
                s0 = peg$FAILED
              }
            } else {
              peg$currPos = s0
              s0 = peg$FAILED
            }
          } else {
            peg$currPos = s0
            s0 = peg$FAILED
          }
        } else {
          peg$currPos = s0
          s0 = peg$FAILED
        }
      } else {
        peg$currPos = s0
        s0 = peg$FAILED
      }
    } else {
      peg$currPos = s0
      s0 = peg$FAILED
    }

    return s0
  }

  function peg$parsecomputeTrafo() {
    var s0, s1, s2, s3, s4, s5, s6, s7, s8, s9

    s0 = peg$currPos
    s1 = peg$parseOPEN()
    if (s1 !== peg$FAILED) {
      s2 = peg$parseo()
      if (s2 !== peg$FAILED) {
        s3 = peg$parsecomputeExpr()
        if (s3 !== peg$FAILED) {
          s4 = []
          s5 = peg$currPos
          s6 = peg$parseo()
          if (s6 !== peg$FAILED) {
            s7 = peg$parseCOMMA()
            if (s7 !== peg$FAILED) {
              s8 = peg$parseo()
              if (s8 !== peg$FAILED) {
                s9 = peg$parsecomputeExpr()
                if (s9 !== peg$FAILED) {
                  s6 = [s6, s7, s8, s9]
                  s5 = s6
                } else {
                  peg$currPos = s5
                  s5 = peg$FAILED
                }
              } else {
                peg$currPos = s5
                s5 = peg$FAILED
              }
            } else {
              peg$currPos = s5
              s5 = peg$FAILED
            }
          } else {
            peg$currPos = s5
            s5 = peg$FAILED
          }
          while (s5 !== peg$FAILED) {
            s4.push(s5)
            s5 = peg$currPos
            s6 = peg$parseo()
            if (s6 !== peg$FAILED) {
              s7 = peg$parseCOMMA()
              if (s7 !== peg$FAILED) {
                s8 = peg$parseo()
                if (s8 !== peg$FAILED) {
                  s9 = peg$parsecomputeExpr()
                  if (s9 !== peg$FAILED) {
                    s6 = [s6, s7, s8, s9]
                    s5 = s6
                  } else {
                    peg$currPos = s5
                    s5 = peg$FAILED
                  }
                } else {
                  peg$currPos = s5
                  s5 = peg$FAILED
                }
              } else {
                peg$currPos = s5
                s5 = peg$FAILED
              }
            } else {
              peg$currPos = s5
              s5 = peg$FAILED
            }
          }
          if (s4 !== peg$FAILED) {
            s5 = peg$parseo()
            if (s5 !== peg$FAILED) {
              s6 = peg$parseCLOSE()
              if (s6 !== peg$FAILED) {
                s1 = [s1, s2, s3, s4, s5, s6]
                s0 = s1
              } else {
                peg$currPos = s0
                s0 = peg$FAILED
              }
            } else {
              peg$currPos = s0
              s0 = peg$FAILED
            }
          } else {
            peg$currPos = s0
            s0 = peg$FAILED
          }
        } else {
          peg$currPos = s0
          s0 = peg$FAILED
        }
      } else {
        peg$currPos = s0
        s0 = peg$FAILED
      }
    } else {
      peg$currPos = s0
      s0 = peg$FAILED
    }

    return s0
  }

  function peg$parsecomputeExpr() {
    var s0, s1, s2

    s0 = peg$currPos
    s1 = peg$parsewhere_clause()
    if (s1 !== peg$FAILED) {
      s2 = peg$parseasAlias()
      if (s2 !== peg$FAILED) {
        s1 = [s1, s2]
        s0 = s1
      } else {
        peg$currPos = s0
        s0 = peg$FAILED
      }
    } else {
      peg$currPos = s0
      s0 = peg$FAILED
    }

    return s0
  }

  function peg$parsecommonFuncTrafo() {
    var s0, s1, s2, s3, s4, s5, s6, s7, s8, s9

    s0 = peg$currPos
    s1 = peg$parseOPEN()
    if (s1 !== peg$FAILED) {
      s2 = peg$parseo()
      if (s2 !== peg$FAILED) {
        s3 = peg$parseoperand()
        if (s3 !== peg$FAILED) {
          s4 = peg$parseo()
          if (s4 !== peg$FAILED) {
            s5 = peg$parseCOMMA()
            if (s5 !== peg$FAILED) {
              s6 = peg$parseo()
              if (s6 !== peg$FAILED) {
                s7 = peg$parseoperand()
                if (s7 !== peg$FAILED) {
                  s8 = peg$parseo()
                  if (s8 !== peg$FAILED) {
                    s9 = peg$parseCLOSE()
                    if (s9 !== peg$FAILED) {
                      s1 = [s1, s2, s3, s4, s5, s6, s7, s8, s9]
                      s0 = s1
                    } else {
                      peg$currPos = s0
                      s0 = peg$FAILED
                    }
                  } else {
                    peg$currPos = s0
                    s0 = peg$FAILED
                  }
                } else {
                  peg$currPos = s0
                  s0 = peg$FAILED
                }
              } else {
                peg$currPos = s0
                s0 = peg$FAILED
              }
            } else {
              peg$currPos = s0
              s0 = peg$FAILED
            }
          } else {
            peg$currPos = s0
            s0 = peg$FAILED
          }
        } else {
          peg$currPos = s0
          s0 = peg$FAILED
        }
      } else {
        peg$currPos = s0
        s0 = peg$FAILED
      }
    } else {
      peg$currPos = s0
      s0 = peg$FAILED
    }

    return s0
  }

  function peg$parseidentityTrafo() {
    var s0

    if (input.substr(peg$currPos, 8) === peg$c172) {
      s0 = peg$c172
      peg$currPos += 8
    } else {
      s0 = peg$FAILED
      if (peg$silentFails === 0) {
        peg$fail(peg$c173)
      }
    }

    return s0
  }

  function peg$parsestring() {
    var s0, s1, s2, s3, s4

    peg$silentFails++
    s0 = peg$currPos
    if (input.charCodeAt(peg$currPos) === 39) {
      s1 = peg$c175
      peg$currPos++
    } else {
      s1 = peg$FAILED
      if (peg$silentFails === 0) {
        peg$fail(peg$c176)
      }
    }
    if (s1 !== peg$FAILED) {
      s2 = peg$currPos
      s3 = []
      if (input.substr(peg$currPos, 2) === peg$c177) {
        s4 = peg$c177
        peg$currPos += 2
      } else {
        s4 = peg$FAILED
        if (peg$silentFails === 0) {
          peg$fail(peg$c178)
        }
      }
      if (s4 === peg$FAILED) {
        if (peg$c179.test(input.charAt(peg$currPos))) {
          s4 = input.charAt(peg$currPos)
          peg$currPos++
        } else {
          s4 = peg$FAILED
          if (peg$silentFails === 0) {
            peg$fail(peg$c180)
          }
        }
      }
      while (s4 !== peg$FAILED) {
        s3.push(s4)
        if (input.substr(peg$currPos, 2) === peg$c177) {
          s4 = peg$c177
          peg$currPos += 2
        } else {
          s4 = peg$FAILED
          if (peg$silentFails === 0) {
            peg$fail(peg$c178)
          }
        }
        if (s4 === peg$FAILED) {
          if (peg$c179.test(input.charAt(peg$currPos))) {
            s4 = input.charAt(peg$currPos)
            peg$currPos++
          } else {
            s4 = peg$FAILED
            if (peg$silentFails === 0) {
              peg$fail(peg$c180)
            }
          }
        }
      }
      if (s3 !== peg$FAILED) {
        s2 = input.substring(s2, peg$currPos)
      } else {
        s2 = s3
      }
      if (s2 !== peg$FAILED) {
        if (input.charCodeAt(peg$currPos) === 39) {
          s3 = peg$c175
          peg$currPos++
        } else {
          s3 = peg$FAILED
          if (peg$silentFails === 0) {
            peg$fail(peg$c176)
          }
        }
        if (s3 !== peg$FAILED) {
          peg$savedPos = s0
          s1 = peg$c181(s2)
          s0 = s1
        } else {
          peg$currPos = s0
          s0 = peg$FAILED
        }
      } else {
        peg$currPos = s0
        s0 = peg$FAILED
      }
    } else {
      peg$currPos = s0
      s0 = peg$FAILED
    }
    peg$silentFails--
    if (s0 === peg$FAILED) {
      s1 = peg$FAILED
      if (peg$silentFails === 0) {
        peg$fail(peg$c174)
      }
    }

    return s0
  }

  function peg$parsenumber() {
    var s0, s1, s2, s3, s4, s5, s6, s7, s8, s9

    s0 = peg$currPos
    s1 = peg$currPos
    s2 = peg$currPos
    if (peg$c182.test(input.charAt(peg$currPos))) {
      s3 = input.charAt(peg$currPos)
      peg$currPos++
    } else {
      s3 = peg$FAILED
      if (peg$silentFails === 0) {
        peg$fail(peg$c183)
      }
    }
    if (s3 === peg$FAILED) {
      s3 = null
    }
    if (s3 !== peg$FAILED) {
      s4 = []
      if (peg$c184.test(input.charAt(peg$currPos))) {
        s5 = input.charAt(peg$currPos)
        peg$currPos++
      } else {
        s5 = peg$FAILED
        if (peg$silentFails === 0) {
          peg$fail(peg$c185)
        }
      }
      if (s5 !== peg$FAILED) {
        while (s5 !== peg$FAILED) {
          s4.push(s5)
          if (peg$c184.test(input.charAt(peg$currPos))) {
            s5 = input.charAt(peg$currPos)
            peg$currPos++
          } else {
            s5 = peg$FAILED
            if (peg$silentFails === 0) {
              peg$fail(peg$c185)
            }
          }
        }
      } else {
        s4 = peg$FAILED
      }
      if (s4 !== peg$FAILED) {
        s5 = peg$currPos
        if (input.charCodeAt(peg$currPos) === 46) {
          s6 = peg$c186
          peg$currPos++
        } else {
          s6 = peg$FAILED
          if (peg$silentFails === 0) {
            peg$fail(peg$c187)
          }
        }
        if (s6 !== peg$FAILED) {
          s7 = []
          if (peg$c184.test(input.charAt(peg$currPos))) {
            s8 = input.charAt(peg$currPos)
            peg$currPos++
          } else {
            s8 = peg$FAILED
            if (peg$silentFails === 0) {
              peg$fail(peg$c185)
            }
          }
          if (s8 !== peg$FAILED) {
            while (s8 !== peg$FAILED) {
              s7.push(s8)
              if (peg$c184.test(input.charAt(peg$currPos))) {
                s8 = input.charAt(peg$currPos)
                peg$currPos++
              } else {
                s8 = peg$FAILED
                if (peg$silentFails === 0) {
                  peg$fail(peg$c185)
                }
              }
            }
          } else {
            s7 = peg$FAILED
          }
          if (s7 !== peg$FAILED) {
            s6 = [s6, s7]
            s5 = s6
          } else {
            peg$currPos = s5
            s5 = peg$FAILED
          }
        } else {
          peg$currPos = s5
          s5 = peg$FAILED
        }
        if (s5 === peg$FAILED) {
          s5 = null
        }
        if (s5 !== peg$FAILED) {
          s6 = peg$currPos
          if (input.charCodeAt(peg$currPos) === 101) {
            s7 = peg$c188
            peg$currPos++
          } else {
            s7 = peg$FAILED
            if (peg$silentFails === 0) {
              peg$fail(peg$c189)
            }
          }
          if (s7 !== peg$FAILED) {
            s8 = []
            if (peg$c184.test(input.charAt(peg$currPos))) {
              s9 = input.charAt(peg$currPos)
              peg$currPos++
            } else {
              s9 = peg$FAILED
              if (peg$silentFails === 0) {
                peg$fail(peg$c185)
              }
            }
            if (s9 !== peg$FAILED) {
              while (s9 !== peg$FAILED) {
                s8.push(s9)
                if (peg$c184.test(input.charAt(peg$currPos))) {
                  s9 = input.charAt(peg$currPos)
                  peg$currPos++
                } else {
                  s9 = peg$FAILED
                  if (peg$silentFails === 0) {
                    peg$fail(peg$c185)
                  }
                }
              }
            } else {
              s8 = peg$FAILED
            }
            if (s8 !== peg$FAILED) {
              s7 = [s7, s8]
              s6 = s7
            } else {
              peg$currPos = s6
              s6 = peg$FAILED
            }
          } else {
            peg$currPos = s6
            s6 = peg$FAILED
          }
          if (s6 === peg$FAILED) {
            s6 = null
          }
          if (s6 !== peg$FAILED) {
            s3 = [s3, s4, s5, s6]
            s2 = s3
          } else {
            peg$currPos = s2
            s2 = peg$FAILED
          }
        } else {
          peg$currPos = s2
          s2 = peg$FAILED
        }
      } else {
        peg$currPos = s2
        s2 = peg$FAILED
      }
    } else {
      peg$currPos = s2
      s2 = peg$FAILED
    }
    if (s2 !== peg$FAILED) {
      s1 = input.substring(s1, peg$currPos)
    } else {
      s1 = s2
    }
    if (s1 !== peg$FAILED) {
      peg$savedPos = s0
      s1 = peg$c190(s1)
    }
    s0 = s1

    return s0
  }

  function peg$parseinteger() {
    var s0, s1, s2, s3, s4, s5

    s0 = peg$currPos
    s1 = peg$currPos
    s2 = peg$currPos
    if (peg$c182.test(input.charAt(peg$currPos))) {
      s3 = input.charAt(peg$currPos)
      peg$currPos++
    } else {
      s3 = peg$FAILED
      if (peg$silentFails === 0) {
        peg$fail(peg$c183)
      }
    }
    if (s3 === peg$FAILED) {
      s3 = null
    }
    if (s3 !== peg$FAILED) {
      s4 = []
      if (peg$c184.test(input.charAt(peg$currPos))) {
        s5 = input.charAt(peg$currPos)
        peg$currPos++
      } else {
        s5 = peg$FAILED
        if (peg$silentFails === 0) {
          peg$fail(peg$c185)
        }
      }
      if (s5 !== peg$FAILED) {
        while (s5 !== peg$FAILED) {
          s4.push(s5)
          if (peg$c184.test(input.charAt(peg$currPos))) {
            s5 = input.charAt(peg$currPos)
            peg$currPos++
          } else {
            s5 = peg$FAILED
            if (peg$silentFails === 0) {
              peg$fail(peg$c185)
            }
          }
        }
      } else {
        s4 = peg$FAILED
      }
      if (s4 !== peg$FAILED) {
        s3 = [s3, s4]
        s2 = s3
      } else {
        peg$currPos = s2
        s2 = peg$FAILED
      }
    } else {
      peg$currPos = s2
      s2 = peg$FAILED
    }
    if (s2 !== peg$FAILED) {
      s1 = input.substring(s1, peg$currPos)
    } else {
      s1 = s2
    }
    if (s1 !== peg$FAILED) {
      peg$savedPos = s0
      s1 = peg$c191(s1)
    }
    s0 = s1

    return s0
  }

  function peg$parseidentifier() {
    var s0, s1, s2, s3, s4

    s0 = peg$currPos
    s1 = peg$currPos
    if (peg$c71.test(input.charAt(peg$currPos))) {
      s2 = input.charAt(peg$currPos)
      peg$currPos++
    } else {
      s2 = peg$FAILED
      if (peg$silentFails === 0) {
        peg$fail(peg$c72)
      }
    }
    if (s2 !== peg$FAILED) {
      s3 = []
      if (peg$c192.test(input.charAt(peg$currPos))) {
        s4 = input.charAt(peg$currPos)
        peg$currPos++
      } else {
        s4 = peg$FAILED
        if (peg$silentFails === 0) {
          peg$fail(peg$c193)
        }
      }
      while (s4 !== peg$FAILED) {
        s3.push(s4)
        if (peg$c192.test(input.charAt(peg$currPos))) {
          s4 = input.charAt(peg$currPos)
          peg$currPos++
        } else {
          s4 = peg$FAILED
          if (peg$silentFails === 0) {
            peg$fail(peg$c193)
          }
        }
      }
      if (s3 !== peg$FAILED) {
        s2 = [s2, s3]
        s1 = s2
      } else {
        peg$currPos = s1
        s1 = peg$FAILED
      }
    } else {
      peg$currPos = s1
      s1 = peg$FAILED
    }
    if (s1 !== peg$FAILED) {
      s0 = input.substring(s0, peg$currPos)
    } else {
      s0 = s1
    }

    return s0
  }

  function peg$parseguid() {
    var s0, s1, s2, s3, s4, s5, s6, s7, s8

    s0 = peg$currPos
    s1 = peg$currPos
    s2 = peg$currPos
    s3 = []
    if (peg$c194.test(input.charAt(peg$currPos))) {
      s4 = input.charAt(peg$currPos)
      peg$currPos++
    } else {
      s4 = peg$FAILED
      if (peg$silentFails === 0) {
        peg$fail(peg$c195)
      }
    }
    if (s4 !== peg$FAILED) {
      while (s4 !== peg$FAILED) {
        s3.push(s4)
        if (peg$c194.test(input.charAt(peg$currPos))) {
          s4 = input.charAt(peg$currPos)
          peg$currPos++
        } else {
          s4 = peg$FAILED
          if (peg$silentFails === 0) {
            peg$fail(peg$c195)
          }
        }
      }
    } else {
      s3 = peg$FAILED
    }
    if (s3 !== peg$FAILED) {
      if (input.charCodeAt(peg$currPos) === 45) {
        s4 = peg$c196
        peg$currPos++
      } else {
        s4 = peg$FAILED
        if (peg$silentFails === 0) {
          peg$fail(peg$c197)
        }
      }
      if (s4 !== peg$FAILED) {
        s5 = []
        s6 = peg$currPos
        s7 = []
        if (peg$c194.test(input.charAt(peg$currPos))) {
          s8 = input.charAt(peg$currPos)
          peg$currPos++
        } else {
          s8 = peg$FAILED
          if (peg$silentFails === 0) {
            peg$fail(peg$c195)
          }
        }
        if (s8 !== peg$FAILED) {
          while (s8 !== peg$FAILED) {
            s7.push(s8)
            if (peg$c194.test(input.charAt(peg$currPos))) {
              s8 = input.charAt(peg$currPos)
              peg$currPos++
            } else {
              s8 = peg$FAILED
              if (peg$silentFails === 0) {
                peg$fail(peg$c195)
              }
            }
          }
        } else {
          s7 = peg$FAILED
        }
        if (s7 !== peg$FAILED) {
          if (input.charCodeAt(peg$currPos) === 45) {
            s8 = peg$c196
            peg$currPos++
          } else {
            s8 = peg$FAILED
            if (peg$silentFails === 0) {
              peg$fail(peg$c197)
            }
          }
          if (s8 === peg$FAILED) {
            s8 = null
          }
          if (s8 !== peg$FAILED) {
            s7 = [s7, s8]
            s6 = s7
          } else {
            peg$currPos = s6
            s6 = peg$FAILED
          }
        } else {
          peg$currPos = s6
          s6 = peg$FAILED
        }
        if (s6 !== peg$FAILED) {
          while (s6 !== peg$FAILED) {
            s5.push(s6)
            s6 = peg$currPos
            s7 = []
            if (peg$c194.test(input.charAt(peg$currPos))) {
              s8 = input.charAt(peg$currPos)
              peg$currPos++
            } else {
              s8 = peg$FAILED
              if (peg$silentFails === 0) {
                peg$fail(peg$c195)
              }
            }
            if (s8 !== peg$FAILED) {
              while (s8 !== peg$FAILED) {
                s7.push(s8)
                if (peg$c194.test(input.charAt(peg$currPos))) {
                  s8 = input.charAt(peg$currPos)
                  peg$currPos++
                } else {
                  s8 = peg$FAILED
                  if (peg$silentFails === 0) {
                    peg$fail(peg$c195)
                  }
                }
              }
            } else {
              s7 = peg$FAILED
            }
            if (s7 !== peg$FAILED) {
              if (input.charCodeAt(peg$currPos) === 45) {
                s8 = peg$c196
                peg$currPos++
              } else {
                s8 = peg$FAILED
                if (peg$silentFails === 0) {
                  peg$fail(peg$c197)
                }
              }
              if (s8 === peg$FAILED) {
                s8 = null
              }
              if (s8 !== peg$FAILED) {
                s7 = [s7, s8]
                s6 = s7
              } else {
                peg$currPos = s6
                s6 = peg$FAILED
              }
            } else {
              peg$currPos = s6
              s6 = peg$FAILED
            }
          }
        } else {
          s5 = peg$FAILED
        }
        if (s5 !== peg$FAILED) {
          s3 = [s3, s4, s5]
          s2 = s3
        } else {
          peg$currPos = s2
          s2 = peg$FAILED
        }
      } else {
        peg$currPos = s2
        s2 = peg$FAILED
      }
    } else {
      peg$currPos = s2
      s2 = peg$FAILED
    }
    if (s2 !== peg$FAILED) {
      s1 = input.substring(s1, peg$currPos)
    } else {
      s1 = s2
    }
    if (s1 !== peg$FAILED) {
      peg$savedPos = s0
      s1 = peg$c198(s1)
    }
    s0 = s1

    return s0
  }

  function peg$parseCOLON() {
    var s0, s1, s2, s3

    s0 = peg$currPos
    s1 = peg$parseo()
    if (s1 !== peg$FAILED) {
      if (input.charCodeAt(peg$currPos) === 58) {
        s2 = peg$c199
        peg$currPos++
      } else {
        s2 = peg$FAILED
        if (peg$silentFails === 0) {
          peg$fail(peg$c200)
        }
      }
      if (s2 !== peg$FAILED) {
        s3 = peg$parseo()
        if (s3 !== peg$FAILED) {
          s1 = [s1, s2, s3]
          s0 = s1
        } else {
          peg$currPos = s0
          s0 = peg$FAILED
        }
      } else {
        peg$currPos = s0
        s0 = peg$FAILED
      }
    } else {
      peg$currPos = s0
      s0 = peg$FAILED
    }

    return s0
  }

  function peg$parseCOMMA() {
    var s0, s1, s2, s3

    s0 = peg$currPos
    s1 = peg$parseo()
    if (s1 !== peg$FAILED) {
      if (input.charCodeAt(peg$currPos) === 44) {
        s2 = peg$c201
        peg$currPos++
      } else {
        s2 = peg$FAILED
        if (peg$silentFails === 0) {
          peg$fail(peg$c202)
        }
      }
      if (s2 !== peg$FAILED) {
        s3 = peg$parseo()
        if (s3 !== peg$FAILED) {
          s1 = [s1, s2, s3]
          s0 = s1
        } else {
          peg$currPos = s0
          s0 = peg$FAILED
        }
      } else {
        peg$currPos = s0
        s0 = peg$FAILED
      }
    } else {
      peg$currPos = s0
      s0 = peg$FAILED
    }

    return s0
  }

  function peg$parseSEMI() {
    var s0, s1, s2, s3

    s0 = peg$currPos
    s1 = peg$parseo()
    if (s1 !== peg$FAILED) {
      if (input.charCodeAt(peg$currPos) === 59) {
        s2 = peg$c45
        peg$currPos++
      } else {
        s2 = peg$FAILED
        if (peg$silentFails === 0) {
          peg$fail(peg$c46)
        }
      }
      if (s2 !== peg$FAILED) {
        s3 = peg$parseo()
        if (s3 !== peg$FAILED) {
          s1 = [s1, s2, s3]
          s0 = s1
        } else {
          peg$currPos = s0
          s0 = peg$FAILED
        }
      } else {
        peg$currPos = s0
        s0 = peg$FAILED
      }
    } else {
      peg$currPos = s0
      s0 = peg$FAILED
    }

    return s0
  }

  function peg$parseOPEN() {
    var s0, s1, s2, s3

    s0 = peg$currPos
    s1 = peg$parseo()
    if (s1 !== peg$FAILED) {
      if (input.charCodeAt(peg$currPos) === 40) {
        s2 = peg$c203
        peg$currPos++
      } else {
        s2 = peg$FAILED
        if (peg$silentFails === 0) {
          peg$fail(peg$c204)
        }
      }
      if (s2 !== peg$FAILED) {
        s3 = peg$parseo()
        if (s3 !== peg$FAILED) {
          s1 = [s1, s2, s3]
          s0 = s1
        } else {
          peg$currPos = s0
          s0 = peg$FAILED
        }
      } else {
        peg$currPos = s0
        s0 = peg$FAILED
      }
    } else {
      peg$currPos = s0
      s0 = peg$FAILED
    }

    return s0
  }

  function peg$parseCLOSE() {
    var s0, s1, s2

    s0 = peg$currPos
    s1 = peg$parseo()
    if (s1 !== peg$FAILED) {
      if (input.charCodeAt(peg$currPos) === 41) {
        s2 = peg$c205
        peg$currPos++
      } else {
        s2 = peg$FAILED
        if (peg$silentFails === 0) {
          peg$fail(peg$c206)
        }
      }
      if (s2 !== peg$FAILED) {
        s1 = [s1, s2]
        s0 = s1
      } else {
        peg$currPos = s0
        s0 = peg$FAILED
      }
    } else {
      peg$currPos = s0
      s0 = peg$FAILED
    }

    return s0
  }

  function peg$parseo() {
    var s0, s1, s2

    peg$silentFails++
    s0 = peg$currPos
    s1 = []
    if (peg$c208.test(input.charAt(peg$currPos))) {
      s2 = input.charAt(peg$currPos)
      peg$currPos++
    } else {
      s2 = peg$FAILED
      if (peg$silentFails === 0) {
        peg$fail(peg$c209)
      }
    }
    while (s2 !== peg$FAILED) {
      s1.push(s2)
      if (peg$c208.test(input.charAt(peg$currPos))) {
        s2 = input.charAt(peg$currPos)
        peg$currPos++
      } else {
        s2 = peg$FAILED
        if (peg$silentFails === 0) {
          peg$fail(peg$c209)
        }
      }
    }
    if (s1 !== peg$FAILED) {
      s0 = input.substring(s0, peg$currPos)
    } else {
      s0 = s1
    }
    peg$silentFails--
    if (s0 === peg$FAILED) {
      s1 = peg$FAILED
      if (peg$silentFails === 0) {
        peg$fail(peg$c207)
      }
    }

    return s0
  }

  function peg$parse_() {
    var s0, s1, s2

    peg$silentFails++
    s0 = peg$currPos
    s1 = []
    if (peg$c208.test(input.charAt(peg$currPos))) {
      s2 = input.charAt(peg$currPos)
      peg$currPos++
    } else {
      s2 = peg$FAILED
      if (peg$silentFails === 0) {
        peg$fail(peg$c209)
      }
    }
    if (s2 !== peg$FAILED) {
      while (s2 !== peg$FAILED) {
        s1.push(s2)
        if (peg$c208.test(input.charAt(peg$currPos))) {
          s2 = input.charAt(peg$currPos)
          peg$currPos++
        } else {
          s2 = peg$FAILED
          if (peg$silentFails === 0) {
            peg$fail(peg$c209)
          }
        }
      }
    } else {
      s1 = peg$FAILED
    }
    if (s1 !== peg$FAILED) {
      s0 = input.substring(s0, peg$currPos)
    } else {
      s0 = s1
    }
    peg$silentFails--
    if (s0 === peg$FAILED) {
      s1 = peg$FAILED
      if (peg$silentFails === 0) {
        peg$fail(peg$c210)
      }
    }

    return s0
  }

  const $ = Object.assign
  // we keep that here to allow for usage in https://pegjs.org/online
  const safeNumber =
    options.safeNumber ||
    function (str) {
      const n = Number(str)
      return Number.isSafeInteger(n) ? n : str
    }
  const hasNotAny$select = columns => {
    return columns.some(cur => cur.expand)
  }
  const { strict } = options
  const stack = []
  let SELECT

  peg$result = peg$startRuleFunction()

  if (peg$result !== peg$FAILED && peg$currPos === input.length) {
    return peg$result
  } else {
    if (peg$result !== peg$FAILED && peg$currPos < input.length) {
      peg$fail(peg$endExpectation())
    }

    throw peg$buildStructuredError(
      peg$maxFailExpected,
      peg$maxFailPos < input.length ? input.charAt(peg$maxFailPos) : null,
      peg$maxFailPos < input.length
        ? peg$computeLocation(peg$maxFailPos, peg$maxFailPos + 1)
        : peg$computeLocation(peg$maxFailPos, peg$maxFailPos)
    )
  }
}

module.exports = {
  SyntaxError: peg$SyntaxError,
  parse: peg$parse
}
