// @sap-cloud-sdk/core/dist/connectivity/scp-cf/client-credentials-token-cache.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.clientCredentialsTokenCache = exports.getGrantTokenCacheKey = void 0;
var moment_1 = __importDefault(require("moment"));
var cache_1 = require("./cache");
var xsuaa_service_1 = require("./xsuaa-service");
var ClientCredentialsTokenCache = function (cache) { return ({
    // TODO: this method name can be shortened
    getGrantTokenFromCache: function (url, credentials) {
        return cache.get(getGrantTokenCacheKey(url, credentials));
    },
    // TODO: this method name can be shortened
    cacheRetrievedToken: function (url, credentials, token) {
        cache.set(getGrantTokenCacheKey(url, credentials), token, token.expires_in
            ? moment_1.default().add(token.expires_in, 'second').unix() * 1000
            : undefined);
    },
    clear: function () {
        cache.clear();
    },
    getCacheInstance: function () { return cache; }
}); };
function getGrantTokenCacheKey(url, credentials) {
    return [url, xsuaa_service_1.headerForClientCredentials(credentials).substring(6)].join(':');
}
exports.getGrantTokenCacheKey = getGrantTokenCacheKey;
exports.clientCredentialsTokenCache = ClientCredentialsTokenCache(new cache_1.Cache());
