// @sap/cds/lib/env/requires.js
const _runtime = '@sap/cds-runtime/lib'

module.exports = {
  "db": undefined,
  "multitenancy": undefined,
  "app-service": {
    // this is the default implementation used for provided services
    impl: `${_runtime}/cds-services/services/Service.js`
  },
  "auth": {
    '[development]': { kind: 'mocked-auth' },
    '[production]': { kind: 'JWT-auth' }
  },
  "dummy-auth": {
    strategy: 'dummy',
  },
  "mocked-auth": {
    strategy: 'mock',
    users: {
      alice: { roles: ['admin'] },
      bob: { roles: ['builder'] },
      '*': true
    }
  },
  "JWT-auth": {
    strategy: 'JWT',
  },
  destinations: undefined,
  monitoring: undefined,
  logging: undefined,
  audit: undefined,
  "sql": {
    '[development]': { kind: 'sqlite', credentials: { database: ':memory:' } },
    '[production]': { kind: 'hana' },
  },
  "sqlite": {
    // use: '@sap/cds-sqlite', // enforce using sqlite when accessed through <sql>
    use: 'sqlite', // enforce using sqlite when accessed through <sql>
    credentials: { database: 'sqlite.db' },
    impl: `${_runtime}/sqlite/Service.js`
  },
  "hana": {
    // use: '@sap/cds-hana', // enforce using hana when accessed through <sql>
    use: 'hana', // enforce using hana when accessed through <sql>
    impl: `${_runtime}/hana/Service.js`
  },
  "rest": {
    impl: `${_runtime}/remote/Service.js`
  },
  "odata": {
    impl: `${_runtime}/remote/Service.js`
  },
  "local-messaging": {
    impl: `${_runtime}/messaging/service.js`
  },
  "file-based-messaging": {
    outbox: true,
    impl: `${_runtime}/messaging/file-based.js`,
    credentials: { file:'~/.cds-msg-box' }
  },
  "enterprise-messaging-shared": {
    outbox: true,
    impl: `${_runtime}/messaging/enterprise-messaging-shared.js`,
    vcap: {
      label: "enterprise-messaging"
    }
  },
  "enterprise-messaging": {
    outbox: true,
    impl: `${_runtime}/messaging/enterprise-messaging.js`
  },
  'message-queuing': {
    outbox: true,
    impl: `${_runtime}/messaging/message-queuing.js`
  },
  "composite-messaging": {
    impl: `${_runtime}/messaging/composite.js`
  }
}
