// @sap/cds-compiler/lib/modelCompare/compare.js
'use strict';

const {
  forEachDefinition,
  forEachMember,
  hasBoolAnnotation
} = require('../model/csnUtils');

/**
 * Compares two models, in HANA-transformed CSN format, to each other.
 *
 * @param beforeModel the before-model
 * @param afterModel the after-model
 * @returns {object} the sets of deletions, extensions, and migrations of entities necessary to transform the before-model
 * to the after-model, together with all the definitions of the after-model
 */
function compareModels(beforeModel, afterModel) {
  const deletedEntities = Object.create(null);
  const elementAdditions = [];
  const elementChanges = [];

  // There is currently no use in knowing the added entities only. If this changes, hand in `addedEntities` to `getArtifactComparator` below.
  forEachDefinition(afterModel, getArtifactComparator(beforeModel, null, null, elementAdditions, elementChanges));       // (, alerts(afterModel))
  forEachDefinition(beforeModel, getArtifactComparator(afterModel, null, deletedEntities, null, null));       // (, alerts(beforeModel))

  const returnObj = Object.create(null);
  returnObj.definitions = afterModel.definitions;
  returnObj.deletions = deletedEntities;
  returnObj.extensions = elementAdditions;
  returnObj.migrations = elementChanges;
  return returnObj;
}

function getArtifactComparator(otherModel, addedEntities, deletedEntities, elementAdditions, elementChanges) {                              // (, alerts)
  return function compareArtifacts(artifact, name) {                                   // (, topKey, path)   topKey == 'definitions'
    function addElements() {
      const elements = {};
      forEachMember(artifact, getElementComparator(otherArtifact, elements));
      if (Object.keys(elements).length > 0) {
        elementAdditions.push({
          extend: name,
          elements: elements
        });
      }
    }
    function removeOrChangeElements() {
      const removedElements = {};
      const changedElements = {};
      const modification = { migrate: name };

      forEachMember(otherArtifact, getElementComparator(artifact, removedElements));
      if (Object.keys(removedElements).length > 0) {
        modification.remove = removedElements;
      }

      forEachMember(artifact, getElementComparator(otherArtifact, null, changedElements));
      if (Object.keys(changedElements).length > 0) {
        modification.change = changedElements;
      }

      if (modification.remove || modification.change) {
        elementChanges.push(modification);
      }
    }

    const otherArtifact = otherModel.definitions[name];
    const isPersisted = isPersistedAsTable(artifact);
    const isPersistedOther = otherArtifact && isPersistedAsTable(otherArtifact);

    if (deletedEntities) {
      // Looking for deleted entities only.
      // Arguments are interchanged in this case: `artifact` from beforeModel and `otherArtifact` from afterModel.
      if (isPersisted && !isPersistedOther) {
        deletedEntities[name] = artifact;
      }
      return;
    }

    // Looking for added entities and added/deleted/changed elements.
    // Parameters: `artifact` from afterModel and `otherArtifact` from beforeModel.

    if (!isPersisted) {
      // Artifact not persisted in afterModel.
      return;
    }

    if (!isPersistedOther) {
      if (addedEntities) {
        addedEntities[name] = artifact;
      }
      return;
    }

    // Artifact changed?

    if (elementAdditions) {
      addElements();
    }
    if (elementChanges) {
      removeOrChangeElements();
    }
  };
}

function isPersistedAsTable(artifact) {
  return artifact.kind === 'entity'
      && !artifact._ignore
      && !artifact.abstract
      && (!artifact.query && !artifact.projection || hasBoolAnnotation(artifact, '@cds.persistence.table'))
      && !hasBoolAnnotation(artifact, '@cds.persistence.skip')
      && !hasBoolAnnotation(artifact, '@cds.persistence.exists');
}

function getElementComparator(otherArtifact, addedElements = null, changedElements = null) {                    // (, alerts)
  return function compareElements(element, name) {                                     // (, topKey, path)   topKey == 'elements'
    if (element._ignore) {
      return;
    }

    const otherElement = otherArtifact.elements[name];
    if (otherElement && !otherElement._ignore) {
      // Element type changed?
      if (!changedElements) {
        return;
      }
      if (relevantTypeChange(element.type, otherElement.type) || typeParametersChanged(element, otherElement)) {
        // Type or parameters, e.g. association target, changed.
        changedElements[name] = changedElement(element, otherElement);
      }

      return;
    }

    if (addedElements) {
      addedElements[name] = element;
    }
  }
}

function relevantTypeChange(type, otherType) {
  return otherType !== type && ![type, otherType].every(t => ['cds.Association', 'cds.Composition'].includes(t));
}

/**
 * Returns whether two things are deeply equal.
 * Function-type things are compared in terms of identity,
 * object-type things in terms of deep equality of all of their properties,
 * all other things in terms of strict equality (===).
 *
 * @param a {any} first thing
 * @param b {any} second thing
 * @param include {function} function of a key and a depth, returning true if and only if the given key at the given depth is to be included in comparison
 * @param depth {number} the current depth in property hierarchy below each of the original arguments (positive, counting from 0; don't set)
 * @returns {boolean}
 */
function deepEqual(a, b, include = () => true, depth = 0) {
  function isObject(x) {
    return x !== null && typeof x === 'object';
  }
  function samePropertyCount() {
    return Object.keys(a).length === Object.keys(b).length;
  }
  function allPropertiesEqual() {
    return Object.keys(a).reduce((prev, key) => prev && (!include(key, depth) || deepEqual(a[key], b[key], include, depth + 1)), true);
  }

  return isObject(a)
      ? isObject(b)
          ? samePropertyCount() && allPropertiesEqual()
          : false
      : a === b;
}

/**
 * Returns whether any type parameters differ between two given elements. Ignores whether types themselves differ (`type` property).
 * @param element {object} an element
 * @param otherElement {object} another element
 * @returns {boolean}
 */
function typeParametersChanged(element, otherElement) {
  return !deepEqual(element, otherElement, (key, depth) => !(depth === 0 && key === 'type'));
}

function changedElement(element, otherElement) {
  return {
    old: otherElement,
    new: element
  };
}

module.exports = {
  compareModels,
  deepEqual
};
